{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d27c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = True\n",
    "#if test:\n",
    "#    print('Warning! Developer mode activated')\n",
    "#    input_fasta = '/home/jpereira/OEs/Results/OE1/NamSeqs/Data/seed_extension/all_97_97/blocks.fasta'\n",
    "#    input_fasta = '/home/jpereira/OEs/Results/OE1/NamSeqs/Data/seed_extension/all_97_97_slice_Mp_B/blocks.fasta'\n",
    "#    \n",
    "#    input_fasta = '/home/jpereira/OEs/Results/OE1/NamSeqs/Data/seed_extension/kmer30/blocks.fasta'\n",
    "#    param_min_instances = 0 # Must be always 0 when you are working with extended seeds\n",
    "#    param_threads = 20\n",
    "#    params_cluster_identity = 0.85\n",
    "#    params_cluster_coverage = 0.85\n",
    "#    params_select_internals_alns = False # Must be always false when you are working with extended seeds\n",
    "#    params_retrive_sinlge_aln = False\n",
    "#    \n",
    "#    output_dir = Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/define_blocks/')\n",
    "#    #work_name = 'single_95_95'\n",
    "#    work_name = 'all_95_95'\n",
    "#    work_name = 'all_93_93'\n",
    "#    work_name = 'all_97_97'\n",
    "#    work_name = 'all_97_97_slice_Mp_B'\n",
    "#    work_name = 'all_97_97_slice_Mp_B.2'\n",
    "#    work_name = 'all_97_97_s70'\n",
    "#    work_name = 'kmer30'\n",
    "#    #work_name = 'NamBlocks_trim'\n",
    "#else:\n",
    "#    parser = argparse.ArgumentParser()\n",
    "#    parser.add_argument('--input-fasta')\n",
    "#    parser.add_argument('--input-clust-names', type=float, default=95)\n",
    "#    parser.add_argument('--input-inner-fasta-tsv', type=int, default=50)\n",
    "#    parser.add_argument('--param-threads', type=int, default=20)\n",
    "#    parser.add_argument('--output-blocks-fasta')\n",
    "#    \n",
    "#    # Capture the parsed arguments in a variable (args)\n",
    "#    args = parser.parse_args()\n",
    "#    \n",
    "#    input_fasta = args.input_fasta\n",
    "#    input_inner_fasta_tsv = args.input_inner_fasta_tsv\n",
    "#    param_threads = args.param_threads\n",
    "#    output_blocks_fasta = args.output_blocks_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "731a480f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 09/19/2025 14:16:54\n",
      "New DB name:   /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/blast_db/infasta_db/blocks.fasta\n",
      "New DB title:  /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/seed_extension/toxo_pasteur2/blocks.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 137 sequences in 0.00140905 seconds.\n",
      "\n",
      "\n",
      "makeblastdb ran successfully.\n",
      "Running: blastn -query /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/seed_extension/toxo_pasteur2/blocks.fasta -db /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/blast_db/infasta_db/blocks.fasta -out /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/blastn.infasta_reciprocal.tsv -num_threads 100 -reward 2 -gapextend 2 -gapopen 4 -penalty -2 -word_size 15 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen'\n",
      "blastn ran successfully.\n",
      "Org_size: 1032\n",
      "Selected pident 90%: 0.9825581395348837\n"
     ]
    }
   ],
   "source": [
    "############################################ Loading libraries and user arguments ###########################################\n",
    "\n",
    "import os\n",
    "from utils.blast_utils import unnoise_coords, makeblast_db, blastn, alignment_absolute_start_end, select_internal_aligns, retrieve_single_alignments\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import argparse\n",
    "from Bio import SeqIO\n",
    "from pathlib import Path  \n",
    "\n",
    "\n",
    "set_trim_nam = {\n",
    "    'input_fasta' : '/home/jpereira/OEs/Blocks_Namasivayam.fa',\n",
    "    'param_min_instances' : 0, # Must be always 0 when you are working with extended seeds\n",
    "    'param_threads' : 20,\n",
    "    'params_cluster_identity' : 0.85,\n",
    "    'params_cluster_coverage' : 0.85,\n",
    "    'params_select_internals_alns' : False, # Must be always false when you are working with extended seeds\n",
    "    'params_retrive_sinlge_aln' : False,\n",
    "    'output_dir' : Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/define_blocks/')\n",
    "    }\n",
    "\n",
    "set_auto_bl_nam = {\n",
    "    'input_fasta' : '/home/jpereira/OEs/Results/OE1/NamSeqs/Data/seed_extension/kmer30/blocks.fasta',\n",
    "    'param_min_instances' : 0, # Must be always 0 when you are working with extended seeds\n",
    "    'param_threads' : 20,\n",
    "    'params_cluster_identity' : 0.85,\n",
    "    'params_cluster_coverage' : 0.85,\n",
    "    'params_select_internals_alns' : False, # Must be always false when you are working with extended seeds\n",
    "    'params_retrive_sinlge_aln' : False,\n",
    "    'output_dir' : Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/define_blocks/')\n",
    "    }\n",
    "\n",
    "set_auto_bl_toxop1 = {\n",
    "    'input_fasta' : '/home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/seed_extension/toxo_pasteur1/blocks.fasta',\n",
    "    'param_min_instances' : 0, # Must be always 0 when you are working with extended seeds\n",
    "    'param_threads' : 20,\n",
    "    'params_cluster_identity' : 0.85,\n",
    "    'params_cluster_coverage' : 0.85,\n",
    "    'params_select_internals_alns' : False, # Must be always false when you are working with extended seeds\n",
    "    'params_retrive_sinlge_aln' : False,\n",
    "    'output_dir' : Path('/home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/')\n",
    "    }\n",
    "\n",
    "set_auto_bl_toxop2 = {\n",
    "    'input_fasta' : '/home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/seed_extension/toxo_pasteur2/blocks.fasta',\n",
    "    'param_min_instances' : 0, # Must be always 0 when you are working with extended seeds\n",
    "    'param_threads' : 20,\n",
    "    'params_cluster_identity' : 0.85,\n",
    "    'params_cluster_coverage' : 0.85,\n",
    "    'params_select_internals_alns' : False, # Must be always false when you are working with extended seeds\n",
    "    'params_retrive_sinlge_aln' : False,\n",
    "    'output_dir' : Path('/home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/')\n",
    "    }\n",
    "\n",
    "\n",
    "\n",
    "settings_dict = {\n",
    "    'trim_nam' : set_trim_nam,\n",
    "    'auto_bl_nam' : set_auto_bl_nam,\n",
    "    'auto_bl_toxop1': set_auto_bl_toxop1, \n",
    "    'auto_bl_toxop2': set_auto_bl_toxop2 \n",
    "}\n",
    "\n",
    "work_name = 'auto_bl_toxop2'\n",
    "\n",
    "input_fasta = settings_dict[work_name]['input_fasta']# '/home/jpereira/OEs/Blocks_Namasivayam.fa',\n",
    "param_min_instances = settings_dict[work_name]['param_min_instances']# 0, # Must be always 0 when you are working with extended seeds\n",
    "param_threads = settings_dict[work_name]['param_threads']# 20,\n",
    "params_cluster_identity = settings_dict[work_name]['params_cluster_identity']# 0.85,\n",
    "params_cluster_coverage = settings_dict[work_name]['params_cluster_coverage']# 0.85,\n",
    "params_select_internals_alns = settings_dict[work_name]['params_select_internals_alns']# False, # Must be always false when you are working with extended seeds\n",
    "params_retrive_sinlge_aln = settings_dict[work_name]['params_retrive_sinlge_aln']# False,\n",
    "output_dir = settings_dict[work_name]['output_dir'] #Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/define_blocks/')\n",
    "\n",
    "\n",
    "if work_name:\n",
    "    output_dir = output_dir / work_name    \n",
    "\n",
    "output_dir.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "##### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
    "\n",
    "output_reciprocal_blast_tsv = output_dir / \"blastn.infasta_reciprocal.tsv\"\n",
    "infasta_db_dir = os.path.join(os.path.dirname(output_reciprocal_blast_tsv), 'blast_db', 'infasta_db') #'/home/jpereira/OEs/OE1.v2/Data/blocks/blast_db/infasta_db' \n",
    "os.makedirs(infasta_db_dir, exist_ok=True)\n",
    "\n",
    "## Make a reciprocal blast using blastn function from blast utils\n",
    "infasta_db = makeblast_db(seqs_path=input_fasta, db_out=infasta_db_dir, remove_old_db=True)\n",
    "blastn(blast_input_seqs=input_fasta, blast_db_file=infasta_db, blast_output_table_tsv=output_reciprocal_blast_tsv,\n",
    "             num_threads=100, reward=2, gap_extend=2, gapopen=4, penalty=-2, word_size=15)\n",
    "\n",
    "\n",
    "############################################ Preparing sequences for iterative block finding ###########################################\n",
    "\n",
    "blast_df = pd.read_csv(output_reciprocal_blast_tsv, sep='\\t')\n",
    "blast_df.columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', \n",
    "                    'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore', 'qlen', 'slen']\n",
    "\n",
    "blast_size = blast_df.shape[0]\n",
    "print(f'Org_size: {blast_size}')\n",
    "s0_df = blast_df[(blast_df['pident'] > 90) & (blast_df['length'] > 15)]\n",
    "print(f'Selected pident 90%: {s0_df.shape[0]/blast_size}')\n",
    "s1_df = alignment_absolute_start_end(s0_df)\n",
    "s2_df = select_internal_aligns(df=s1_df,border=10) \n",
    "\n",
    "# Filter by query alignment interval\n",
    "if params_select_internals_alns:\n",
    "    s3_df = s2_df# blast.fuse_contained_intervals(df=s2_df, start_col='qstart', end_col='qend', group_cols=['qseqid', 'sseqid'], keep_highest_score_col='bitscore')\n",
    "else:\n",
    "    s3_df = s1_df\n",
    "\n",
    "##### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5535729b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def get_fasta_from_bed(fasta_file, bed_file, output_file):\n",
    "    # Read the FASTA file into a dictionary for quick access\n",
    "    fasta_dict = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    \n",
    "    with open(bed_file, 'r') as bed, open(output_file, 'w') as out:\n",
    "        for line in bed:\n",
    "            if line.strip() == \"\" or line.startswith(\"#\"):\n",
    "                continue  # Skip empty lines and comments\n",
    "\n",
    "            chrom, start, end = line.strip().split()[:3]\n",
    "            start, end = int(start), int(end)\n",
    "            \n",
    "            if chrom not in fasta_dict:\n",
    "                raise ValueError(f\"Sequence {chrom} not found in FASTA file.\")\n",
    "            \n",
    "            seq = fasta_dict[chrom].seq[start-1:end]\n",
    "            out.write(f\">{chrom}:{start}-{end}\\n{seq}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "408e483e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration dir: /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations\n",
      "Removieng previous iteration dir\n",
      "\n",
      "###### ITERATION 0 ######\n",
      "\n",
      "Retrieving all blocks iteration: 0\n",
      "Number of Retrived Seeds in iteration 0: 235\n",
      "Extracted regions from bed file: /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/regions.fasta\n",
      "Executing command: vsearch --derep_fulllength /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/regions.fasta --output /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/derep.fasta --sizeout --relabel Seq --minseqlength 20\n",
      "Executing command: vsearch --sortbysize /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/derep.fasta --output /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/sorted.fasta\n",
      "Executing command: vsearch --cluster_size /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/sorted.fasta --id 0.85 --query_cov 0.85 --target_cov 0.85 --minseqlength 20 --centroids /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/centroids.fasta --uc /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/clusters.uc --strand both --sizein\n",
      "\n",
      "###### ITERATION 1 ######\n",
      "\n",
      "makeblastdb ran successfully.\n",
      "Running: blastn -query /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/1/centroids.fasta -db /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/blast_db/centroids.fasta -out /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/blastn.reciprocal.tsv -num_threads 80 -reward 1 -gapextend 5 -gapopen 5 -penalty -1 -word_size 10 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen'\n",
      "blastn ran successfully.\n",
      "Retrieving all blocks iteration: 1\n",
      "Number of Retrived Seeds in iteration 1: 43\n",
      "Extracted regions from bed file: /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/regions.fasta\n",
      "Executing command: vsearch --derep_fulllength /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/regions.fasta --output /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/derep.fasta --sizeout --relabel Seq --minseqlength 20\n",
      "Executing command: vsearch --sortbysize /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/derep.fasta --output /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/sorted.fasta\n",
      "Executing command: vsearch --cluster_size /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/sorted.fasta --id 0.85 --query_cov 0.85 --target_cov 0.85 --minseqlength 20 --centroids /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/centroids.fasta --uc /home/jpereira/OEs/Results/OE1/ToxoPasteur1/Data/define_blocks/auto_bl_toxop2/block_iterations/2/clusters.uc --strand both --sizein\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "############################################ Iterative Block finding ###########################################\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import shutil\n",
    "import pandas as pd\n",
    "from utils.bedtools_utils import bedtools_getfasta, coords_to_regions, blast_to_bed\n",
    "from utils.vsearch_utils import vsearch_dereplication, vsearch_clustersize, vsearch_sortbysize\n",
    "\n",
    "\n",
    "iteration_dir = output_dir /  'block_iterations'\n",
    "print(f\"Iteration dir: {iteration_dir}\")\n",
    "if os.path.isdir(iteration_dir): \n",
    "    print(\"Removieng previous iteration dir\")\n",
    "    shutil.rmtree(iteration_dir)\n",
    "iteration_dir.mkdir(exist_ok=True)\n",
    "\n",
    "iteration = 0\n",
    "num_blocks = []\n",
    "while iteration < 2:\n",
    "    \n",
    "    print(f\"\\n###### ITERATION {iteration} ######\\n\")\n",
    "\n",
    "    in_fasta                         = iteration_dir / f'{iteration}'     / 'centroids.fasta'\n",
    "    blast_db                         = iteration_dir / f'{iteration + 1}' / 'blast_db'\n",
    "    reciprocal_blast_tsv             = iteration_dir / f'{iteration + 1}' / 'blastn.reciprocal.tsv'\n",
    "    coords_bed                       = iteration_dir / f'{iteration + 1}' / 'regions.bed'\n",
    "    extracted_regions_fasta          = iteration_dir / f'{iteration + 1}' / 'regions.fasta'\n",
    "    retrieve_coords_bed              = iteration_dir / f'{iteration}'     / 'retrieve' / 'regions.bed'\n",
    "    retrieve_extracted_regions_fasta = iteration_dir / f'{iteration}'     / 'retrieve' / 'regions.fasta'\n",
    "    derep_fasta                      = iteration_dir / f'{iteration + 1}' / 'derep.fasta'\n",
    "    sorted_fasta                     = iteration_dir / f'{iteration + 1}' / 'sorted.fasta'\n",
    "    centroids_fasta                  = iteration_dir / f'{iteration + 1}' / 'centroids.fasta'\n",
    "    clusters_uc                      = iteration_dir / f'{iteration + 1}' / 'clusters.uc'\n",
    "    \n",
    "    iteration_path = os.path.dirname(clusters_uc)\n",
    "    iteration_log =  os.path.join(iteration_path, 'logs')\n",
    "    \n",
    "    (clusters_uc.parent).mkdir( exist_ok=True)\n",
    "    (retrieve_coords_bed.parent).mkdir(exist_ok=True, parents=True)\n",
    "    \n",
    "    ## Make a reciprocal blast using blastn function from blast utils\n",
    "    if iteration == 0:\n",
    "        blast_df = s3_df\n",
    "        in_fasta = input_fasta\n",
    "    else:\n",
    "        blast_db = makeblast_db(seqs_path=in_fasta, db_out=blast_db, remove_old_db=True, log_file=iteration_log)\n",
    "        \n",
    "        blastn(blast_input_seqs=in_fasta,\n",
    "               blast_db_file=blast_db,\n",
    "               blast_output_table_tsv=reciprocal_blast_tsv,\n",
    "               num_threads=80, reward=1, gap_extend=5, gapopen=5,\n",
    "               penalty=-1, word_size=10, log_file=iteration_log)\n",
    "        \n",
    "        blast_df = pd.read_csv(reciprocal_blast_tsv, sep='\\t', header=None)\n",
    "        blast_df.columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', \n",
    "                    'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore', 'qlen', 'slen']\n",
    "        \n",
    "    if iteration == 0:\n",
    "        blast_df0 = blast_df.copy()\n",
    "\n",
    "    blast_df = blast_df[blast_df['pident'] > 90]\n",
    "    blast_df = blast_df[blast_df['length'] > 20]\n",
    "    \n",
    "    if params_retrive_sinlge_aln:\n",
    "        # Retrive aligmentes that only map with itself\n",
    "        retrive_df, blast_df = retrieve_single_alignments(blast_df)\n",
    "        if retrive_df.shape[0] != 0:\n",
    "            print(f\"Retrieving sinlge mapped blocks iteration: {iteration}\")\n",
    "            retrive_bed_df = blast_to_bed(retrive_df)\n",
    "            retrive_bed_df = retrive_bed_df.drop_duplicates()\n",
    "            retrive_bed_df.to_csv(retrieve_coords_bed, sep='\\t', header=False, index=False)\n",
    "            #bedtools_getfasta(fasta_path=in_fasta, bed_path=retrieve_coords_bed, extracted_regions_path=retrieve_extracted_regions_fasta, show_command=True, log_file=iteration_log)\n",
    "            get_fasta_from_bed(in_fasta, retrieve_coords_bed, retrieve_extracted_regions_fasta)\n",
    "    else:\n",
    "        print(f\"Retrieving all blocks iteration: {iteration}\")\n",
    "        retrive_bed_df = blast_to_bed(blast_df)\n",
    "        retrive_bed_df = retrive_bed_df.drop_duplicates()\n",
    "        retrive_bed_df.to_csv(retrieve_coords_bed, sep='\\t', header=False, index=False)\n",
    "        #bedtools_getfasta(fasta_path=in_fasta, bed_path=retrieve_coords_bed, extracted_regions_path=retrieve_extracted_regions_fasta, show_command=True, log_file=iteration_log)\n",
    "        get_fasta_from_bed(in_fasta, retrieve_coords_bed, retrieve_extracted_regions_fasta)\n",
    "    \n",
    "    print(f\"Number of Retrived Seeds in iteration {iteration}: {len(retrive_bed_df)}\")\n",
    "\n",
    "        \n",
    "    if blast_df.empty:\n",
    "        print(\"Descomposition finished, all blocks are independent\")\n",
    "        break\n",
    "\n",
    "    unnoised_df = blast_df.groupby('qseqid').apply(lambda x: unnoise_coords(x, radious=10))\n",
    "    unnoised_df = unnoised_df.reset_index().drop(columns='level_1')\n",
    "\n",
    "    # Filter coordinates with low number of instances \n",
    "    unnoised_df = unnoised_df[unnoised_df['instances'] > param_min_instances]\n",
    "    \n",
    "    if iteration == 0:\n",
    "        unnoised_df0 = unnoised_df.copy()\n",
    "\n",
    "\n",
    "    # Apply the function across groups\n",
    "    get_fasta_df = unnoised_df.groupby('qseqid', group_keys=False).apply(coords_to_regions)\n",
    "    get_fasta_df.to_csv(coords_bed, sep='\\t', header=False, index=False)\n",
    "    #bedtools_getfasta(fasta_path=in_fasta, bed_path=coords_bed, extracted_regions_path=extracted_regions_fasta, show_command=True, log_file=iteration_log)\n",
    "    get_fasta_from_bed(in_fasta, coords_bed, extracted_regions_fasta)\n",
    "    print(f\"Extracted regions from bed file: {extracted_regions_fasta}\")\n",
    "    vsearch_dereplication(input_fasta=extracted_regions_fasta, derep_fasta=derep_fasta, min_seq_length=20, log_file=iteration_log)\n",
    "    vsearch_sortbysize(derep_fasta=derep_fasta, sorted_fasta=sorted_fasta, log_file=iteration_log)\n",
    "    vsearch_clustersize( sorted_fasta=sorted_fasta, centroids_fasta=centroids_fasta, uc_file=clusters_uc,query_cov=params_cluster_coverage,\n",
    "                        target_cov=params_cluster_coverage, id_thresh=params_cluster_identity, min_seq_length=20, log_file=iteration_log,\n",
    "                        use_both_strands=True,use_sizein=True)\n",
    "    \n",
    "    \n",
    "    iteration += 1\n",
    "\n",
    "##### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6ee664e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     Seq5;size=1\n",
       "0     Seq1;size=2\n",
       "0    Seq19;size=1\n",
       "0     Seq7;size=1\n",
       "0    Seq12;size=1\n",
       "Name: id, dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "725395b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq5;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1;size=2</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq19;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq7;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq12;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id  start  end\n",
       "0   Seq5;size=1      1   40\n",
       "0   Seq1;size=2      1   50\n",
       "0  Seq19;size=1      1   65\n",
       "0   Seq7;size=1      1   82\n",
       "0  Seq12;size=1      1   85"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qlen</th>\n",
       "      <th>slen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Seq5;size=1</td>\n",
       "      <td>Seq5;size=1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>3.320000e-15</td>\n",
       "      <td>65.0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1;size=2</td>\n",
       "      <td>Seq1;size=2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>8.060000e-20</td>\n",
       "      <td>80.8</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Seq19;size=1</td>\n",
       "      <td>Seq19;size=1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>7.960000e-27</td>\n",
       "      <td>104.0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Seq7;size=1</td>\n",
       "      <td>Seq7;size=1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>8.230000e-35</td>\n",
       "      <td>131.0</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Seq20;size=1</td>\n",
       "      <td>Seq20;size=1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "      <td>3.190000e-36</td>\n",
       "      <td>136.0</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           qseqid        sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
       "91    Seq5;size=1   Seq5;size=1   100.0      40         0        0       1   \n",
       "0     Seq1;size=2   Seq1;size=2   100.0      50         0        0       1   \n",
       "54   Seq19;size=1  Seq19;size=1   100.0      65         0        0       1   \n",
       "108   Seq7;size=1   Seq7;size=1   100.0      82         0        0       1   \n",
       "57   Seq20;size=1  Seq20;size=1   100.0      85         0        0       1   \n",
       "\n",
       "     qend  sstart  send        evalue  bitscore  qlen  slen  \n",
       "91     40       1    40  3.320000e-15      65.0    40    40  \n",
       "0      50       1    50  8.060000e-20      80.8    50    50  \n",
       "54     65       1    65  7.960000e-27     104.0    65    65  \n",
       "108    82       1    82  8.230000e-35     131.0    82    82  \n",
       "57     85       1    85  3.190000e-36     136.0    85    85  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>coords</th>\n",
       "      <th>instances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq12;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq12;size=1</td>\n",
       "      <td>85</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq19;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq19;size=1</td>\n",
       "      <td>65</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Seq1;size=2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Seq1;size=2</td>\n",
       "      <td>50</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Seq5;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Seq5;size=1</td>\n",
       "      <td>40</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Seq7;size=1</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Seq7;size=1</td>\n",
       "      <td>82</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          qseqid  coords  instances\n",
       "4   Seq12;size=1       1        2.0\n",
       "5   Seq12;size=1      85        2.0\n",
       "18  Seq19;size=1       1        2.0\n",
       "19  Seq19;size=1      65        2.0\n",
       "20   Seq1;size=2       1        2.0\n",
       "21   Seq1;size=2      50        2.0\n",
       "38   Seq5;size=1       1        2.0\n",
       "39   Seq5;size=1      40        2.0\n",
       "42   Seq7;size=1       1        2.0\n",
       "43   Seq7;size=1      82        2.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "first_ids = get_fasta_df.sort_values('end').head()['id']\n",
    "display(get_fasta_df.sort_values('end').head())\n",
    "display(blast_df.sort_values('qlen').head())\n",
    "display(unnoised_df[unnoised_df['qseqid'].isin(first_ids)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ef49b0e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>coords</th>\n",
       "      <th>instances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>block_25</td>\n",
       "      <td>742</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>block_23</td>\n",
       "      <td>781</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>block_19</td>\n",
       "      <td>881</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>block_25</td>\n",
       "      <td>886</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>block_19</td>\n",
       "      <td>965</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>block_25</td>\n",
       "      <td>970</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>block_23</td>\n",
       "      <td>977</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>block_27</td>\n",
       "      <td>1235</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>block_27</td>\n",
       "      <td>1440</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>block_27</td>\n",
       "      <td>1601</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      qseqid  coords  instances\n",
       "71  block_25     742        2.0\n",
       "57  block_23     781        4.0\n",
       "36  block_19     881       13.0\n",
       "68  block_25     886       11.0\n",
       "35  block_19     965       15.0\n",
       "67  block_25     970       15.0\n",
       "56  block_23     977        8.0\n",
       "79  block_27    1235        2.0\n",
       "78  block_27    1440        5.0\n",
       "77  block_27    1601        9.0"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unnoised_df0.sort_values('coords').tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cd60ed6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>block_1</td>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>block_10</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>block_10</td>\n",
       "      <td>70</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>block_10</td>\n",
       "      <td>89</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>block_11</td>\n",
       "      <td>1</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>block_8</td>\n",
       "      <td>162</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>block_8</td>\n",
       "      <td>173</td>\n",
       "      <td>306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>block_8</td>\n",
       "      <td>306</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>block_8</td>\n",
       "      <td>323</td>\n",
       "      <td>405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>block_9</td>\n",
       "      <td>1</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id  start  end\n",
       "0    block_1      1   85\n",
       "0   block_10      1   70\n",
       "1   block_10     70   89\n",
       "2   block_10     89  171\n",
       "0   block_11      1  145\n",
       "..       ...    ...  ...\n",
       "1    block_8    162  173\n",
       "2    block_8    173  306\n",
       "3    block_8    306  323\n",
       "4    block_8    323  405\n",
       "0    block_9      1   91\n",
       "\n",
       "[75 rows x 3 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unnoised_df0['qseqid'].unique())\n",
    "unnoised_df0.groupby('qseqid', group_keys=False).apply(coords_to_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8b580301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/define_blocks/all_97_97_slice_Mp_B.2/block_iterations/2/regions.fasta')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extracted_regions_fasta"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "OE1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
