{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Develper mode activated!\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import argparse\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "\n",
    "test = True\n",
    "if test:\n",
    "    print('Develper mode activated!')\n",
    "    inputs = {\n",
    "        'fasta_in' : '/home/jpereira/OEs/OE1.v2/Data/filter_sizes/toxo1.size_filtered.fasta',\n",
    "        'paf_in'   : '/home/jpereira/OEs/OE1.v2/Data/map_contamination/query_x_ref-contamination.paf'}\n",
    "    outputs = {\n",
    "        'fasta_out' : '/home/jpereira/OEs/OE1.v2/Data/decontamination/test-decont.fasta',\n",
    "        'stats_out' : '/home/jpereira/OEs/OE1.v2/Data/decontamination/test-decont-stats.tsv'}\n",
    "    params = {\n",
    "        'min_coverage' : 70}\n",
    "else:\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description=\"\"\"\n",
    "        Discard sequences from a FASTA file if they align well (≥ min_coverage) \n",
    "        to any reference in the given PAF alignment file.\n",
    "        \"\"\")\n",
    "    parser.add_argument(\"--fasta_in\", help=\"Path to the input (filtered) FASTA file.\")\n",
    "    parser.add_argument(\"--paf_in\", help=\"Path to the PAF alignment file.\")\n",
    "    parser.add_argument(\"--fasta_out\", help=\"Path to the output (decontaminated) FASTA file.\")\n",
    "    parser.add_argument(\"--stats_out\", help=\"Path to the output TSV stats file.\")\n",
    "    parser.add_argument(\"--min_coverage\",type=float,default=70.0,\n",
    "        help=\"Minimum coverage percentage (0-100) to consider a sequence contaminated. Default: 70.\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    inputs = {\n",
    "        'fasta_in' : args.fasta_in,\n",
    "        'paf_in'   : args.paf_in}\n",
    "\n",
    "    outputs = {\n",
    "        'fasta_out' : args.fasta_out,\n",
    "        'stats_out' : args.stats_out}\n",
    "\n",
    "    params = {\n",
    "        'min_coverage' : args.min_coverage}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6d27c6ce-4a8f-4419-9a0f-0350eedbc36a</td>\n",
       "      <td>2192</td>\n",
       "      <td>1424</td>\n",
       "      <td>2129</td>\n",
       "      <td>+</td>\n",
       "      <td>NW_023666047.1</td>\n",
       "      <td>90413263</td>\n",
       "      <td>46636679</td>\n",
       "      <td>46637394</td>\n",
       "      <td>71</td>\n",
       "      <td>715</td>\n",
       "      <td>3</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:5</td>\n",
       "      <td>s1:i:68</td>\n",
       "      <td>s2:i:43</td>\n",
       "      <td>dv:f:0.0727</td>\n",
       "      <td>rl:i:1457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6d27c6ce-4a8f-4419-9a0f-0350eedbc36a</td>\n",
       "      <td>2192</td>\n",
       "      <td>422</td>\n",
       "      <td>1561</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_023666044.1</td>\n",
       "      <td>54000376</td>\n",
       "      <td>53905312</td>\n",
       "      <td>53906446</td>\n",
       "      <td>63</td>\n",
       "      <td>1145</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:4</td>\n",
       "      <td>s1:i:58</td>\n",
       "      <td>s2:i:40</td>\n",
       "      <td>dv:f:0.1065</td>\n",
       "      <td>rl:i:1457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a0994257-2f39-4d85-991a-3eb74523d4a7</td>\n",
       "      <td>9948</td>\n",
       "      <td>1296</td>\n",
       "      <td>6570</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_023666047.1</td>\n",
       "      <td>90413263</td>\n",
       "      <td>46707556</td>\n",
       "      <td>46712888</td>\n",
       "      <td>76</td>\n",
       "      <td>5339</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:4</td>\n",
       "      <td>s1:i:61</td>\n",
       "      <td>s2:i:56</td>\n",
       "      <td>dv:f:0.0919</td>\n",
       "      <td>rl:i:9568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>a0994257-2f39-4d85-991a-3eb74523d4a7</td>\n",
       "      <td>9948</td>\n",
       "      <td>2262</td>\n",
       "      <td>9133</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_023666047.1</td>\n",
       "      <td>90413263</td>\n",
       "      <td>46647778</td>\n",
       "      <td>46654609</td>\n",
       "      <td>65</td>\n",
       "      <td>6871</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:5</td>\n",
       "      <td>s1:i:56</td>\n",
       "      <td>s2:i:53</td>\n",
       "      <td>dv:f:0.0772</td>\n",
       "      <td>rl:i:9568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a0994257-2f39-4d85-991a-3eb74523d4a7</td>\n",
       "      <td>9948</td>\n",
       "      <td>291</td>\n",
       "      <td>1315</td>\n",
       "      <td>+</td>\n",
       "      <td>NW_023666169.1</td>\n",
       "      <td>253443</td>\n",
       "      <td>93810</td>\n",
       "      <td>94852</td>\n",
       "      <td>57</td>\n",
       "      <td>1042</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:3</td>\n",
       "      <td>s1:i:52</td>\n",
       "      <td>s2:i:50</td>\n",
       "      <td>dv:f:0.0704</td>\n",
       "      <td>rl:i:9568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260480</th>\n",
       "      <td>dde7daf8-5312-46bd-90b6-fd8939bfd7a1</td>\n",
       "      <td>3000</td>\n",
       "      <td>44</td>\n",
       "      <td>189</td>\n",
       "      <td>+</td>\n",
       "      <td>CM065578.1</td>\n",
       "      <td>2450754</td>\n",
       "      <td>4544</td>\n",
       "      <td>4689</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>28</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:13</td>\n",
       "      <td>s1:i:145</td>\n",
       "      <td>s2:i:124</td>\n",
       "      <td>dv:f:0.0075</td>\n",
       "      <td>rl:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260481</th>\n",
       "      <td>a40e387b-18b0-4af0-9ed4-58eaa3ca0f9c</td>\n",
       "      <td>1580</td>\n",
       "      <td>265</td>\n",
       "      <td>1569</td>\n",
       "      <td>+</td>\n",
       "      <td>CM065578.1</td>\n",
       "      <td>2450754</td>\n",
       "      <td>18595</td>\n",
       "      <td>19881</td>\n",
       "      <td>1261</td>\n",
       "      <td>1305</td>\n",
       "      <td>33</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:119</td>\n",
       "      <td>s1:i:1256</td>\n",
       "      <td>s2:i:1109</td>\n",
       "      <td>dv:f:0.0030</td>\n",
       "      <td>rl:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260482</th>\n",
       "      <td>a40e387b-18b0-4af0-9ed4-58eaa3ca0f9c</td>\n",
       "      <td>1580</td>\n",
       "      <td>18</td>\n",
       "      <td>367</td>\n",
       "      <td>+</td>\n",
       "      <td>CM065584.1</td>\n",
       "      <td>12128471</td>\n",
       "      <td>3790207</td>\n",
       "      <td>3790556</td>\n",
       "      <td>349</td>\n",
       "      <td>349</td>\n",
       "      <td>6</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:30</td>\n",
       "      <td>s1:i:349</td>\n",
       "      <td>s2:i:340</td>\n",
       "      <td>dv:f:0.0017</td>\n",
       "      <td>rl:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260483</th>\n",
       "      <td>d32d9727-429b-4108-835b-060284048b48</td>\n",
       "      <td>1570</td>\n",
       "      <td>17</td>\n",
       "      <td>1397</td>\n",
       "      <td>+</td>\n",
       "      <td>CM065578.1</td>\n",
       "      <td>2450754</td>\n",
       "      <td>6027</td>\n",
       "      <td>7412</td>\n",
       "      <td>1324</td>\n",
       "      <td>1391</td>\n",
       "      <td>35</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:121</td>\n",
       "      <td>s1:i:1319</td>\n",
       "      <td>s2:i:1155</td>\n",
       "      <td>dv:f:0.0030</td>\n",
       "      <td>rl:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260484</th>\n",
       "      <td>d32d9727-429b-4108-835b-060284048b48</td>\n",
       "      <td>1570</td>\n",
       "      <td>1205</td>\n",
       "      <td>1563</td>\n",
       "      <td>+</td>\n",
       "      <td>CM065576.1</td>\n",
       "      <td>1864358</td>\n",
       "      <td>10</td>\n",
       "      <td>368</td>\n",
       "      <td>358</td>\n",
       "      <td>358</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:31</td>\n",
       "      <td>s1:i:358</td>\n",
       "      <td>s2:i:173</td>\n",
       "      <td>dv:f:0.0017</td>\n",
       "      <td>rl:i:0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>260485 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          0     1     2     3  4   \\\n",
       "0       6d27c6ce-4a8f-4419-9a0f-0350eedbc36a  2192  1424  2129  +   \n",
       "1       6d27c6ce-4a8f-4419-9a0f-0350eedbc36a  2192   422  1561  -   \n",
       "2       a0994257-2f39-4d85-991a-3eb74523d4a7  9948  1296  6570  -   \n",
       "3       a0994257-2f39-4d85-991a-3eb74523d4a7  9948  2262  9133  -   \n",
       "4       a0994257-2f39-4d85-991a-3eb74523d4a7  9948   291  1315  +   \n",
       "...                                      ...   ...   ...   ... ..   \n",
       "260480  dde7daf8-5312-46bd-90b6-fd8939bfd7a1  3000    44   189  +   \n",
       "260481  a40e387b-18b0-4af0-9ed4-58eaa3ca0f9c  1580   265  1569  +   \n",
       "260482  a40e387b-18b0-4af0-9ed4-58eaa3ca0f9c  1580    18   367  +   \n",
       "260483  d32d9727-429b-4108-835b-060284048b48  1570    17  1397  +   \n",
       "260484  d32d9727-429b-4108-835b-060284048b48  1570  1205  1563  +   \n",
       "\n",
       "                    5         6         7         8     9     10  11      12  \\\n",
       "0       NW_023666047.1  90413263  46636679  46637394    71   715   3  tp:A:P   \n",
       "1       NW_023666044.1  54000376  53905312  53906446    63  1145   0  tp:A:P   \n",
       "2       NW_023666047.1  90413263  46707556  46712888    76  5339   0  tp:A:P   \n",
       "3       NW_023666047.1  90413263  46647778  46654609    65  6871   0  tp:A:P   \n",
       "4       NW_023666169.1    253443     93810     94852    57  1042   0  tp:A:P   \n",
       "...                ...       ...       ...       ...   ...   ...  ..     ...   \n",
       "260480      CM065578.1   2450754      4544      4689   145   145  28  tp:A:P   \n",
       "260481      CM065578.1   2450754     18595     19881  1261  1305  33  tp:A:P   \n",
       "260482      CM065584.1  12128471   3790207   3790556   349   349   6  tp:A:P   \n",
       "260483      CM065578.1   2450754      6027      7412  1324  1391  35  tp:A:P   \n",
       "260484      CM065576.1   1864358        10       368   358   358  60  tp:A:P   \n",
       "\n",
       "              13         14         15           16         17  \n",
       "0         cm:i:5    s1:i:68    s2:i:43  dv:f:0.0727  rl:i:1457  \n",
       "1         cm:i:4    s1:i:58    s2:i:40  dv:f:0.1065  rl:i:1457  \n",
       "2         cm:i:4    s1:i:61    s2:i:56  dv:f:0.0919  rl:i:9568  \n",
       "3         cm:i:5    s1:i:56    s2:i:53  dv:f:0.0772  rl:i:9568  \n",
       "4         cm:i:3    s1:i:52    s2:i:50  dv:f:0.0704  rl:i:9568  \n",
       "...          ...        ...        ...          ...        ...  \n",
       "260480   cm:i:13   s1:i:145   s2:i:124  dv:f:0.0075     rl:i:0  \n",
       "260481  cm:i:119  s1:i:1256  s2:i:1109  dv:f:0.0030     rl:i:0  \n",
       "260482   cm:i:30   s1:i:349   s2:i:340  dv:f:0.0017     rl:i:0  \n",
       "260483  cm:i:121  s1:i:1319  s2:i:1155  dv:f:0.0030     rl:i:0  \n",
       "260484   cm:i:31   s1:i:358   s2:i:173  dv:f:0.0017     rl:i:0  \n",
       "\n",
       "[260485 rows x 18 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.read_csv(inputs['paf_in'], sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import argparse\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "\n",
    "test = True\n",
    "if test:\n",
    "    print('Develper mode activated!')\n",
    "    inputs = {\n",
    "        'fasta_in' : '/home/jpereira/OEs/OE1.v2/Data/filter_sizes/toxo1.size_filtered.fasta',\n",
    "        'paf_in'   : '/home/jpereira/OEs/OE1.v2/Data/map_contamination/query_x_ref-contamination.paf'}\n",
    "    outputs = {\n",
    "        'fasta_out' : '/home/jpereira/OEs/OE1.v2/Data/decontamination/test-decont.fasta',\n",
    "        'stats_out' : '/home/jpereira/OEs/OE1.v2/Data/decontamination/test-decont-stats.tsv'}\n",
    "    params = {\n",
    "        'min_coverage' : 70}\n",
    "else:\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description=\"\"\"\n",
    "        Discard sequences from a FASTA file if they align well (≥ min_coverage) \n",
    "        to any reference in the given PAF alignment file.\n",
    "        \"\"\")\n",
    "    parser.add_argument(\"--fasta_in\", help=\"Path to the input (filtered) FASTA file.\")\n",
    "    parser.add_argument(\"--paf_in\", help=\"Path to the PAF alignment file.\")\n",
    "    parser.add_argument(\"--fasta_out\", help=\"Path to the output (decontaminated) FASTA file.\")\n",
    "    parser.add_argument(\"--stats_out\", help=\"Path to the output TSV stats file.\")\n",
    "    parser.add_argument(\"--min_coverage\",type=float,default=70.0,\n",
    "        help=\"Minimum coverage percentage (0-100) to consider a sequence contaminated. Default: 70.\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    inputs = {\n",
    "        'fasta_in' : args.fasta_in,\n",
    "        'paf_in'   : args.paf_in}\n",
    "\n",
    "    outputs = {\n",
    "        'fasta_out' : args.fasta_out,\n",
    "        'stats_out' : args.stats_out}\n",
    "\n",
    "    params = {\n",
    "        'min_coverage' : args.min_coverage}\n",
    "\n",
    "\n",
    "\n",
    "def parse_paf_line(line):\n",
    "    \"\"\"\n",
    "    Parse one line from a PAF file and return:\n",
    "      - qname (query/contig ID)\n",
    "      - qlen (query length)\n",
    "      - qstart (start of the alignment on the query)\n",
    "      - qend (end of the alignment on the query)\n",
    "    The PAF format has many fields, but these four are enough to estimate coverage.\n",
    "    \"\"\"\n",
    "    fields = line.strip().split('\\t')\n",
    "    qname = fields[0]\n",
    "    qlen = int(fields[1])\n",
    "    qstart = int(fields[2])\n",
    "    qend = int(fields[3])\n",
    "    matching_bases = int(fields[9])\n",
    "    return qname, qlen, qstart, qend, matching_bases\n",
    "\n",
    "\n",
    "\n",
    "def main(inputs: dict, outputs: dict, params: dict):\n",
    "\n",
    "    # Make a directory for the output files\n",
    "    os.makedirs(os.path.dirname(outputs['fasta_out']), exist_ok=True)\n",
    "    \n",
    "    # Store the highest coverage found for each query\n",
    "    coverage_dict = {}\n",
    "\n",
    "    # 1. Parse the PAF file to identify contaminated sequences\n",
    "    with open(inputs['paf_in'], 'r') as paf_file:\n",
    "        for line in paf_file:\n",
    "            # Skip empty or comment lines\n",
    "            if not line.strip() or line.startswith('#'):\n",
    "                continue\n",
    "            qname, qlen, qstart, qend, matching_bases = parse_paf_line(line)\n",
    "            \n",
    "            # Calculate coverage: (aligned length / total query length) * 100\n",
    "            aligned_len = abs(qend - qstart)\n",
    "            coverage = (matching_bases / qlen) * 100 if qlen > 0 else 0\n",
    "            \n",
    "            # Keep track of the maximum coverage observed for this sequence\n",
    "            if qname not in coverage_dict or coverage > coverage_dict[qname]:\n",
    "                coverage_dict[qname] = coverage\n",
    "\n",
    "    # 2. Read the FASTA and separate contaminated vs. clean sequences\n",
    "    contaminated = set()\n",
    "    total_sequences = 0\n",
    "\n",
    "    with open(inputs['fasta_in'], 'r') as fasta_file:\n",
    "        # We won't write out here yet, just counting\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            total_sequences += 1\n",
    "            seq_id = record.id\n",
    "            max_cov = coverage_dict.get(seq_id, 0.0)\n",
    "            \n",
    "            # If coverage >= threshold, mark as contaminated\n",
    "            if max_cov >= params['min_coverage']:\n",
    "                contaminated.add(seq_id)\n",
    "\n",
    "    # 3. Write out the decontaminated sequences to a new FASTA\n",
    "    retained_count = 0\n",
    "    with open(inputs['fasta_in'], 'r') as fasta_file, open(outputs['fasta_out'], 'w') as out_fasta:\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            if record.id not in contaminated:\n",
    "                retained_count += 1\n",
    "                SeqIO.write(record, out_fasta, \"fasta\")\n",
    "\n",
    "    # 4. Write stats to TSV\n",
    "    contaminated_count = len(contaminated)\n",
    "    with open(outputs['stats_out'], 'w') as stats_file:\n",
    "        stats_file.write(\"Total\\tContaminated\\tRetained\\tMinCoverage\\n\")\n",
    "        stats_file.write(f\"{total_sequences}\\t{contaminated_count}\\t{retained_count}\\t{params['min_coverage']}\\n\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(inputs, outputs, params)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "OE1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
