{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30bd3327",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "❌ Skipping non-FASTA/Q file: /home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/.snakemake_timestamp\n",
      "⚠️ Skipping cluster_2.fasta: only 0 entries (min is 35)\n",
      "▶ Running: kmercountexact.sh in=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta k=35 threads=5 mincount=35 fastadump=f out=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/cluster_1.tsv overwrite=true\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "java -ea -Xmx170951m -Xms170951m -cp /usr/local/bin/current/ jgi.KmerCountExact in=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta k=35 threads=5 mincount=35 fastadump=f out=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/cluster_1.tsv overwrite=true\n",
      "Executing jgi.KmerCountExact [in=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta, k=35, threads=5, mincount=35, fastadump=f, out=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/cluster_1.tsv, overwrite=true]\n",
      "\n",
      "Set threads to 5\n",
      "Initial:\n",
      "Memory: max=179281m, total=179281m, free=179195m, used=86m\n",
      "\n",
      "Executing ukmer.KmerTableSetU [in=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta, k=35, threads=5, mincount=35, fastadump=f, out=/home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/cluster_1.tsv, overwrite=true]\n",
      "\n",
      "K was changed from 35 to 34\n",
      "Initial:\n",
      "Ways=31, initialSize=128000, prefilter=f, prealloc=f\n",
      "Memory: max=179281m, total=179281m, free=179179m, used=102m\n",
      "\n",
      "After table allocation:\n",
      "Memory: max=179281m, total=179281m, free=179078m, used=203m\n",
      "\n",
      "After loading:\n",
      "Memory: max=179281m, total=179281m, free=175706m, used=3575m\n",
      "\n",
      "Input:                      \t209 reads \t\t629329 bases.\n",
      "\n",
      "For K=34\n",
      "Unique Kmers:               \t60411\n",
      "Average Kmer Count:         \t10.303\n",
      "Estimated Kmer Depth:       \t10.303\n",
      "Estimated Read Depth:       \t10.417\n",
      "\n",
      "Load Time:                  \t0.178 seconds.\n",
      "Kmer Dump Time:             \t0.628 seconds.\n"
     ]
    }
   ],
   "source": [
    "from utils.bedtools_utils import bbtools_kmercountexact\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "import argparse\n",
    "import re\n",
    "\n",
    "\n",
    "test = True\n",
    "if test:\n",
    "    input_fasta_dir = Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/')\n",
    "    params_min_fasta_size = 35\n",
    "    params_kmer_size = 35\n",
    "    params_min_kmer_counts = 35\n",
    "    params_threads = 5\n",
    "    output_seeds_dir = Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/')\n",
    "else:\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('--input-fasta-dir', required=True, help=\"directory with the sequences useed to make seeds\")    \n",
    "    parser.add_argument('--params-min-fasta-size', default=35, help='Minimum number of reads reqeried in each fasta')\n",
    "    parser.add_argument('--params-kmersize', default=30, help='Size of K-mer')\n",
    "    parser.add_argument('--params-threads', default=5)\n",
    "    parser.add_argument('--output-seeds-dir', required=True, help='Output directory with the seeds tables in tsv format')\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    input_fasta_dir        = args.input_fasta_dir\n",
    "    params_min_fasta_size  = args.params_min_fasta_size\n",
    "    params_kmer_size       = args.params_kmer_size\n",
    "    params_min_kmer_counts = args.params_min_kmer_counts\n",
    "    params_threads         = args.params_threads\n",
    "    output_seeds_dir       = args.output_seeds_dir\n",
    "\n",
    "output_seeds_dir.mkdir(exist_ok=True)\n",
    "\n",
    "\n",
    "# Loop over files in the input directory\n",
    "for fasta in input_fasta_dir.iterdir():\n",
    "    \n",
    "    if not fasta.is_file():\n",
    "        continue\n",
    "\n",
    "    if not re.search(r\"\\.(fasta|fa|fastq)$\", fasta.name):\n",
    "        print(f\"❌ Skipping non-FASTA/Q file: {fasta}\")\n",
    "        continue\n",
    "    \n",
    "    # Count the number of entries in the sequence file\n",
    "    q = sum(1 for _ in SeqIO.parse(fasta, \"fasta\"))  # or \"fastq\" depending on format\n",
    "    \n",
    "    if q < params_min_fasta_size:\n",
    "        print(f\"⚠️ Skipping {fasta.name}: only {q} entries (min is {params_min_fasta_size})\")\n",
    "        continue\n",
    "\n",
    "    # Remove file extension using regex\n",
    "    file_name = re.sub(r\"\\.(fasta|fa|fastq)$\", \"\", fasta.name)\n",
    "    \n",
    "    output_seed_tsv = output_seeds_dir / f\"{file_name}.tsv\"\n",
    "    \n",
    "    bbtools_kmercountexact(\n",
    "        infile=fasta,\n",
    "        outfile=output_seed_tsv,\n",
    "        k=params_kmer_size,\n",
    "        min_counts=params_min_kmer_counts,\n",
    "        khist=None,\n",
    "        make_fasta=False,\n",
    "        threads=params_threads,\n",
    "        extra=None,\n",
    "        exe=\"kmercountexact.sh\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c4796d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "627e7452",
   "metadata": {},
   "source": [
    "\n",
    "# Old script used to generate seeds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d9555b23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning! Developer mode activated\n",
      "\n",
      "\n",
      "Building a new DB, current time: 06/27/2025 18:13:50\n",
      "New DB name:   /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/blast_db/infasta_db/cluster_1.fasta\n",
      "New DB title:  /home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 209 sequences in 0.00580883 seconds.\n",
      "\n",
      "\n",
      "makeblastdb ran successfully.\n",
      "Running: blastn -query /home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta -db /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/blast_db/infasta_db/cluster_1.fasta -out /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/blastn.infasta_reciprocal.tsv -num_threads 100 -reward 2 -gapextend 2 -gapopen 4 -penalty -2 -word_size 15 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen'\n",
      "blastn ran successfully.\n",
      "Org_size: 260422\n",
      "Filtered pident 90%: 0.951901145064549\n",
      "Iteration dir: /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations\n",
      "Removieng previous iteration dir\n",
      "\n",
      "###### ITERATION 0 ######\n",
      "\n",
      "Retrieving all blocks iteration: 0\n",
      "Running: bedtools getfasta -fi /home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta -bed /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/0/retrieve/regions.bed -fo /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/0/retrieve/regions.fasta\n",
      "Number of Retrived Seeds in iteration 0: 34656\n",
      "Running: bedtools getfasta -fi /home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta -bed /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/regions.bed -fo /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/regions.fasta\n",
      "Executing command: vsearch --derep_fulllength /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/regions.fasta --output /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/derep.fasta --sizeout --relabel Seq --minseqlength 32\n",
      "Executing command: vsearch --sortbysize /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/derep.fasta --output /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/sorted.fasta\n",
      "Executing command: vsearch --cluster_size /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/sorted.fasta --id 0.97 --query_cov 0.97 --target_cov 0.97 --minseqlength 32 --centroids /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/centroids.fasta --uc /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/clusters.uc --strand both --sizein\n",
      "\n",
      "###### ITERATION 1 ######\n",
      "\n",
      "makeblastdb ran successfully.\n",
      "Running: blastn -query /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/centroids.fasta -db /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/blast_db/centroids.fasta -out /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/blastn.reciprocal.tsv -num_threads 80 -reward 1 -gapextend 5 -gapopen 5 -penalty -1 -word_size 10 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen'\n",
      "blastn ran successfully.\n",
      "Retrieving all blocks iteration: 1\n",
      "Running: bedtools getfasta -fi /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/centroids.fasta -bed /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/retrieve/regions.bed -fo /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/retrieve/regions.fasta\n",
      "Number of Retrived Seeds in iteration 1: 10920\n",
      "Running: bedtools getfasta -fi /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/1/centroids.fasta -bed /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/regions.bed -fo /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/regions.fasta\n",
      "Executing command: vsearch --derep_fulllength /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/regions.fasta --output /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/derep.fasta --sizeout --relabel Seq --minseqlength 32\n",
      "Executing command: vsearch --sortbysize /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/derep.fasta --output /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/sorted.fasta\n",
      "Executing command: vsearch --cluster_size /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/sorted.fasta --id 0.97 --query_cov 0.97 --target_cov 0.97 --minseqlength 32 --centroids /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/centroids.fasta --uc /home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/all_97_97/block_iterations/2/clusters.uc --strand both --sizein\n"
     ]
    }
   ],
   "source": [
    "############################################ Loading libraries and user arguments ###########################################\n",
    "\n",
    "import os\n",
    "from utils.blast_utils import unnoise_coords, makeblast_db, blastn, alignment_absolute_start_end, select_internal_aligns, retrieve_single_alignments\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import argparse\n",
    "from Bio import SeqIO\n",
    "from pathlib import Path  \n",
    "\n",
    "test = True\n",
    "if test:\n",
    "    print('Warning! Developer mode activated')\n",
    "    input_fasta = '/home/jpereira/OEs/Results/OE1/NamSeqs/Data/mcl_clustering/sequneces_clusters/cluster_1.fasta'\n",
    "    param_min_instances = 1\n",
    "    param_threads = 90\n",
    "    params_cluster_identity = 0.97\n",
    "    params_cluster_coverage = 0.97\n",
    "    params_select_internals_alns = True\n",
    "    params_retrive_sinlge_aln = False\n",
    "    \n",
    "    output_dir = Path('/home/jpereira/OEs/Results/OE1/NamSeqs/Data/make_seeds/')\n",
    "    work_name = 'single_95_95'\n",
    "    work_name = 'all_93_93'\n",
    "    work_name = 'all_97_97'\n",
    "     \n",
    "else:\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('--input-fasta')\n",
    "    parser.add_argument('--input-clust-names', type=float, default=95)\n",
    "    parser.add_argument('--input-inner-fasta-tsv', type=int, default=50)\n",
    "    parser.add_argument('--param-threads', type=int, default=20)\n",
    "    parser.add_argument('--output-blocks-fasta')\n",
    "    \n",
    "    # Capture the parsed arguments in a variable (args)\n",
    "    args = parser.parse_args()\n",
    "    \n",
    "    input_fasta = args.input_fasta\n",
    "    input_inner_fasta_tsv = args.input_inner_fasta_tsv\n",
    "    param_threads = args.param_threads\n",
    "    output_blocks_fasta = args.output_blocks_fasta\n",
    "    \n",
    "if work_name:\n",
    "    output_dir = output_dir / work_name    \n",
    "\n",
    "output_dir.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "############################################ Preparing sequences for iterative block finding ###########################################\n",
    "\n",
    "output_reciprocal_blast_tsv = output_dir / \"blastn.infasta_reciprocal.tsv\"\n",
    "infasta_db_dir = os.path.join(os.path.dirname(output_reciprocal_blast_tsv), 'blast_db', 'infasta_db') #'/home/jpereira/OEs/OE1.v2/Data/blocks/blast_db/infasta_db' \n",
    "os.makedirs(infasta_db_dir, exist_ok=True)\n",
    "\n",
    "## Make a reciprocal blast using blastn function from blast utils\n",
    "infasta_db = makeblast_db(seqs_path=input_fasta, db_out=infasta_db_dir, remove_old_db=True)\n",
    "blastn(blast_input_seqs=input_fasta, blast_db_file=infasta_db, blast_output_table_tsv=output_reciprocal_blast_tsv,\n",
    "             num_threads=100, reward=2, gap_extend=2, gapopen=4, penalty=-2, word_size=15)\n",
    "\n",
    "\n",
    "##### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
    "\n",
    "############################################ Preparing sequences for iterative block finding ###########################################\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "blast_df = pd.read_csv(output_reciprocal_blast_tsv, sep='\\t')\n",
    "blast_df.columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', \n",
    "                    'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore', 'qlen', 'slen']\n",
    "\n",
    "blast_size = blast_df.shape[0]\n",
    "print(f'Org_size: {blast_size}')\n",
    "s0_df = blast_df[(blast_df['pident'] > 90) & (blast_df['length'] > 15)]\n",
    "print(f'Filtered pident 90%: {s0_df.shape[0]/blast_size}')\n",
    "s1_df = alignment_absolute_start_end(s0_df)\n",
    "s2_df = select_internal_aligns(df=s1_df,border=10) \n",
    "\n",
    "# Filter by query alignment interval\n",
    "if params_select_internals_alns:\n",
    "    s3_df = s2_df# blast.fuse_contained_intervals(df=s2_df, start_col='qstart', end_col='qend', group_cols=['qseqid', 'sseqid'], keep_highest_score_col='bitscore')\n",
    "else:\n",
    "    s3_df = s1_df\n",
    "\n",
    "##### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
    "\n",
    "\n",
    "############################################ Iterative Block finding ###########################################\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import shutil\n",
    "import pandas as pd\n",
    "from utils.bedtools_utils import bedtools_getfasta, coords_to_regions, blast_to_bed\n",
    "from utils.vsearch_utils import vsearch_dereplication, vsearch_clustersize, vsearch_sortbysize\n",
    "\n",
    "\n",
    "iteration_dir = output_dir /  'block_iterations'\n",
    "print(f\"Iteration dir: {iteration_dir}\")\n",
    "if os.path.isdir(iteration_dir): \n",
    "    print(\"Removieng previous iteration dir\")\n",
    "    shutil.rmtree(iteration_dir)\n",
    "iteration_dir.mkdir(exist_ok=True)\n",
    "\n",
    "iteration = 0\n",
    "num_blocks = []\n",
    "while iteration < 2:\n",
    "    \n",
    "    print(f\"\\n###### ITERATION {iteration} ######\\n\")\n",
    "\n",
    "    in_fasta                         = iteration_dir / f'{iteration}'     / 'centroids.fasta'\n",
    "    blast_db                         = iteration_dir / f'{iteration + 1}' / 'blast_db'\n",
    "    reciprocal_blast_tsv             = iteration_dir / f'{iteration + 1}' / 'blastn.reciprocal.tsv'\n",
    "    coords_bed                       = iteration_dir / f'{iteration + 1}' / 'regions.bed'\n",
    "    extracted_regions_fasta          = iteration_dir / f'{iteration + 1}' / 'regions.fasta'\n",
    "    retrieve_coords_bed              = iteration_dir / f'{iteration}'     / 'retrieve' / 'regions.bed'\n",
    "    retrieve_extracted_regions_fasta = iteration_dir / f'{iteration}'     / 'retrieve' / 'regions.fasta'\n",
    "    derep_fasta                      = iteration_dir / f'{iteration + 1}' / 'derep.fasta'\n",
    "    sorted_fasta                     = iteration_dir / f'{iteration + 1}' / 'sorted.fasta'\n",
    "    centroids_fasta                  = iteration_dir / f'{iteration + 1}' / 'centroids.fasta'\n",
    "    clusters_uc                      = iteration_dir / f'{iteration + 1}' / 'clusters.uc'\n",
    "    \n",
    "    iteration_path = os.path.dirname(clusters_uc)\n",
    "    iteration_log =  os.path.join(iteration_path, 'logs')\n",
    "    \n",
    "    (clusters_uc.parent).mkdir( exist_ok=True)\n",
    "    (retrieve_coords_bed.parent).mkdir(exist_ok=True, parents=True)\n",
    "    \n",
    "    ## Make a reciprocal blast using blastn function from blast utils\n",
    "    if iteration == 0:\n",
    "        blast_df = s3_df\n",
    "        in_fasta = input_fasta\n",
    "    else:\n",
    "        blast_db = makeblast_db(seqs_path=in_fasta, db_out=blast_db, remove_old_db=True, log_file=iteration_log)\n",
    "        \n",
    "        blastn(blast_input_seqs=in_fasta,\n",
    "               blast_db_file=blast_db,\n",
    "               blast_output_table_tsv=reciprocal_blast_tsv,\n",
    "               num_threads=80, reward=1, gap_extend=5, gapopen=5,\n",
    "               penalty=-1, word_size=10, log_file=iteration_log)\n",
    "        \n",
    "        blast_df = pd.read_csv(reciprocal_blast_tsv, sep='\\t')\n",
    "        blast_df.columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', \n",
    "                    'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore', 'qlen', 'slen']\n",
    "\n",
    "    blast_df = blast_df[blast_df['pident'] > 90]\n",
    "    blast_df = blast_df[blast_df['length'] > 20]\n",
    "    \n",
    "    if params_retrive_sinlge_aln:\n",
    "        # Retrive aligmentes that only map with itself\n",
    "        retrive_df, blast_df = retrieve_single_alignments(blast_df)\n",
    "        if retrive_df.shape[0] != 0:\n",
    "            print(f\"Retrieving sinlge mapped blocks iteration: {iteration}\")\n",
    "            retrive_bed_df = blast_to_bed(retrive_df)\n",
    "            retrive_bed_df = retrive_bed_df.drop_duplicates()\n",
    "            retrive_bed_df.to_csv(retrieve_coords_bed, sep='\\t', header=False, index=False)\n",
    "            bedtools_getfasta(fasta_path=in_fasta, bed_path=retrieve_coords_bed, extracted_regions_path=retrieve_extracted_regions_fasta, show_command=True, log_file=iteration_log)\n",
    "    else:\n",
    "        print(f\"Retrieving all blocks iteration: {iteration}\")\n",
    "        retrive_bed_df = blast_to_bed(blast_df)\n",
    "        retrive_bed_df = retrive_bed_df.drop_duplicates()\n",
    "        retrive_bed_df.to_csv(retrieve_coords_bed, sep='\\t', header=False, index=False)\n",
    "        bedtools_getfasta(fasta_path=in_fasta, bed_path=retrieve_coords_bed, extracted_regions_path=retrieve_extracted_regions_fasta, show_command=True, log_file=iteration_log)\n",
    "    \n",
    "    print(f\"Number of Retrived Seeds in iteration {iteration}: {len(retrive_bed_df)}\")\n",
    "\n",
    "        \n",
    "    if blast_df.empty:\n",
    "        print(\"Descomposition finished, all blocks are independent\")\n",
    "        break\n",
    "\n",
    "    unnoised_df = blast_df.groupby('qseqid').apply(lambda x: unnoise_coords(x, radious=15))\n",
    "    unnoised_df = unnoised_df.reset_index().drop(columns='level_1')\n",
    "\n",
    "    # Filter coordinates with low number of instances \n",
    "    unnoised_df = unnoised_df[unnoised_df['instances'] > param_min_instances]\n",
    "\n",
    "    # Apply the function across groups\n",
    "    get_fasta_df = unnoised_df.groupby('qseqid', group_keys=False).apply(coords_to_regions)\n",
    "    get_fasta_df.to_csv(coords_bed, sep='\\t', header=False, index=False)\n",
    "    bedtools_getfasta(fasta_path=in_fasta, bed_path=coords_bed, extracted_regions_path=extracted_regions_fasta, show_command=True, log_file=iteration_log)\n",
    "    vsearch_dereplication(input_fasta=extracted_regions_fasta, derep_fasta=derep_fasta, min_seq_length=32, log_file=iteration_log)\n",
    "    vsearch_sortbysize(derep_fasta=derep_fasta, sorted_fasta=sorted_fasta, log_file=iteration_log)\n",
    "    vsearch_clustersize( sorted_fasta=sorted_fasta, centroids_fasta=centroids_fasta, uc_file=clusters_uc,query_cov=params_cluster_coverage,\n",
    "                        target_cov=params_cluster_coverage, id_thresh=params_cluster_identity, min_seq_length=32, log_file=iteration_log,\n",
    "                        use_both_strands=True,use_sizein=True)\n",
    "    \n",
    "    \n",
    "    iteration += 1\n",
    "\n",
    "##### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "73fba73a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1;size=92</td>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1;size=92</td>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq1;size=92</td>\n",
       "      <td>3</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq1;size=92</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq1;size=92</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33766</th>\n",
       "      <td>Seq988;size=1</td>\n",
       "      <td>44</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33767</th>\n",
       "      <td>Seq988;size=1</td>\n",
       "      <td>50</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33768</th>\n",
       "      <td>Seq988;size=1</td>\n",
       "      <td>44</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33769</th>\n",
       "      <td>Seq988;size=1</td>\n",
       "      <td>52</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33770</th>\n",
       "      <td>Seq988;size=1</td>\n",
       "      <td>44</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10920 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              qseqid  qstart  qend\n",
       "0       Seq1;size=92       1    45\n",
       "8       Seq1;size=92       1    44\n",
       "12      Seq1;size=92       3    45\n",
       "13      Seq1;size=92       1    43\n",
       "16      Seq1;size=92       1    42\n",
       "...              ...     ...   ...\n",
       "33766  Seq988;size=1      44    85\n",
       "33767  Seq988;size=1      50    85\n",
       "33768  Seq988;size=1      44    79\n",
       "33769  Seq988;size=1      52    89\n",
       "33770  Seq988;size=1      44    76\n",
       "\n",
       "[10920 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "retrive_bed_df.drop_duplicates()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "OE1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
