{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scripts Objectives:\n",
    "# -Creata a table were each rows correspond to a different query sequences, \n",
    "# and each column indicates in which references the query sequence was mapped.\n",
    "# -The table will be composed by the follwoing columns:\n",
    "#   query_name, max_coverage, max_base_num_coverage, max_identity*, max_good_qaln_qaln_qaln_qaln_aln_qidentity**, \n",
    "#    Ref1.map_status† Ref1.max_coverage, Ref1.max_base_num_coverage, Ref1.max_identity\n",
    "#    Ref2.map_status†, . . .  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total reads length: 6216\n",
      "Sequence number: 24\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import os, re, sys\n",
    "\n",
    "test = False\n",
    "mito_test = True\n",
    "# Replace 'your_fasta_file.fasta' with the path to your FASTA file\n",
    "if test:\n",
    "    input_query_fasta_path = '../toxo_reads/toxo1.trimmedReads.fasta'\n",
    "    input_ref_fasta_dir_path = '../Data/map_contamination'\n",
    "    output_unmapped_reads_path = '../Data/unmapped_reads.fasta'\n",
    "    output_map_statistics = ''\n",
    "elif mito_test:\n",
    "    input_query_fasta_path = '../Blocks_Namasivayam.fa'\n",
    "    input_ref_fasta_dir_path = '../Data/map_contamination'\n",
    "    output_unmapped_reads_path = '../Data/mapped_blocks_Namasivayam.fasta'\n",
    "    output_map_statistics = ''\n",
    "else:\n",
    "    input_query_fasta_path = sys.argv[1]\n",
    "    input_ref_fasta_dir_path = sys.argv[2]\n",
    "    output_unmapped_reads_path = sys.argv[3]\n",
    "    output_map_statistics = sys.argv[4]\n",
    "\n",
    "# Parse the FASTA file and extract the lengths\n",
    "total_len = 0\n",
    "num_records = 0\n",
    "query_len = []\n",
    "query_id = []\n",
    "for record in SeqIO.parse(input_query_fasta_path, \"fasta\"):\n",
    "    total_len += len(record.seq)\n",
    "    num_records += 1\n",
    "    query_len.append(len(record.seq))\n",
    "    query_id.append(record.id)\n",
    "\n",
    "paf_df = pd.DataFrame({'qname' : query_id, 'qlen': query_len})\n",
    "\n",
    "print(f'Total reads length: {total_len}')\n",
    "print(f'Sequence number: {num_records}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mito_test=False\n",
    "test = True\n",
    "if test:\n",
    "    paf_file1_path = '../Data/map_contamination/TGVEG.paf'\n",
    "    paf_file2_path = '../Data/map_contamination/TGA4.paf'\n",
    "    paf_file3_path = '../Data/map_contamination/TgRH_pasteur.paf'\n",
    "    paf_file4_path = '../Data/map_contamination/Chlorocebus_sabaeus.paf'\n",
    "    paf_file5_path = '../Data/map_contamination/Micoplasma_tauri.paf'\n",
    "    paf_file6_path = '../Data/map_contamination/Homo_sapiens.paf'\n",
    "    paf_file6_path = '../Data/map_contamination/TG_APlast.paf'\n",
    "    paf_file7_path = '../Data/map_contamination/Ncanium_LIV.paf'\n",
    "\n",
    "    paf_files_path_list = [paf_file1_path, paf_file2_path, paf_file3_path, paf_file4_path, paf_file5_path, paf_file6_path, paf_file7_path]\n",
    "elif mito_test:\n",
    "    paf_file1_path = '../Data/map_contamination/TGVEG.mito.paf'\n",
    "    paf_file2_path = '../Data/map_contamination/TGA4.mito.paf'\n",
    "    paf_file3_path = '../Data/map_contamination/TgRH_pasteur.mito.paf'\n",
    "    paf_file4_path = '../Data/map_contamination/Chlorocebus_sabaeus.mito.paf'\n",
    "    paf_file5_path = '../Data/map_contamination/Micoplasma_tauri.mito.paf'\n",
    "    paf_file6_path = '../Data/map_contamination/Homo_sapiens.mito.paf'\n",
    "    paf_file6_path = '../Data/map_contamination/TG_APlast.mito.paf'\n",
    "    paf_file7_path = '../Data/map_contamination/Ncanium_LIV.mito.paf'\n",
    "    paf_files_path_list = [paf_file1_path, paf_file2_path, paf_file3_path, paf_file4_path, paf_file5_path, paf_file6_path, paf_file7_path]\n",
    "\n",
    "else:\n",
    "    paf_files_path_list = os.listdir(input_ref_fasta_dir_path)\n",
    "\n",
    "# Define the path to your PAF file\n",
    "\n",
    "# Define column names based on the PAF specification\n",
    "column_names = [\n",
    "    'qname', 'qlen', 'qstart', 'qend', 'strand', 'sname', 'slen', 'sstart',\n",
    "    'send', 'matching_residues', 'aln_len', 'minimap_ql'\n",
    "    ,'undf1','undf2','undf3','undf4','undf5','undf6']\n",
    "    #'aln_type',\t'minimizers',\t'qscore',\t'sscore',\t'read_len', 'divergence']\n",
    "    # qned - qlen: Shows the number of query bases that were mapped\n",
    "    # aln_len: Shows the number of subject bases that were mapped\n",
    "\n",
    "# Load the PAF file into a DataFrame\n",
    "paf_df = pd.DataFrame({'qname' : query_id, 'qlen' : query_len})\n",
    "for path in paf_files_path_list:\n",
    "\n",
    "    df = pd.read_csv(path, sep='\\t', header=None, names=column_names, index_col=None)\n",
    "    if df.empty:\n",
    "        print(f'{os.path.basename(path)} dataframe is empty')\n",
    "        continue\n",
    "\n",
    "    file_name = os.path.basename(path)\n",
    "    org_name = re.sub('.[^.]+$', '', file_name)\n",
    "    \n",
    "    df['org_name'] = org_name\n",
    "    df['coverage'] =(df['qend'] - df['qstart']).abs() / df['qlen']\n",
    "    df['aln_qidentity'] = df['matching_residues'] / (df['qend'] - df['qstart']).abs()\n",
    "    df['total_qidentity'] = df['matching_residues'] / df['qlen']\n",
    "    df['undefined'] = df[['undf1','undf2','undf3','undf4','undf5','undf6']].astype(str).agg('|'.join, axis=1)\n",
    "    \n",
    "    pattern = r'(dv:f:(0|1)\\.{0,1}[0-9]*)'\n",
    "    df['divergence'] = df['undefined'].str.extract(pattern)[0]\n",
    "    df['divergence'] = df['divergence'].str.replace(r'dv:f:' , '')\n",
    "    df['divergence'] = df['divergence'].astype(float)\n",
    "    \n",
    "    # If a query has more than two alignaments (for the same reference),\n",
    "    # select only the one with the greatest aln_qidentity\n",
    "    max_total_qident = df.groupby('qname')['total_qidentity'].idxmax() \n",
    "    max_aln_qident = df.groupby('qname')['aln_qidentity'].idxmax() \n",
    "    max_qcover = df.groupby('qname')['coverage'].idxmax() \n",
    "    min_divergence = df.groupby('qname')['divergence'].idxmin() \n",
    "\n",
    "    df_total_qident = pd.DataFrame()\n",
    "    df_aln_qident = pd.DataFrame()\n",
    "    df_coverage = pd.DataFrame()\n",
    "    df_divergence = pd.DataFrame()\n",
    "    \n",
    "    df_total_qident[['qname', f'{org_name}.max_total_qidentity']] = df.loc[max_aln_qident.values, ['qname','aln_qidentity']]\n",
    "    df_aln_qident[['qname', f'{org_name}.max_aln_qidentity']] = df.loc[max_aln_qident.values, ['qname','aln_qidentity']]\n",
    "    df_coverage[['qname', f'{org_name}.max_coverage']] = df.loc[max_qcover.values, ['qname','coverage']]\n",
    "    df_divergence[['qname', f'{org_name}.min_divergence']] = df.loc[min_divergence.values, ['qname','divergence']]\n",
    "\n",
    "    df_org = pd.merge(df_aln_qident, df_coverage, on='qname', how='outer')\n",
    "    df_org = pd.merge(df_org, df_divergence, on='qname', how='outer')\n",
    "    df_org = pd.merge(df_org, df_total_qident, on='qname', how='outer')\n",
    "\n",
    "    paf_df = pd.merge(paf_df, df_org, on='qname', how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_total_qident_cols = []\n",
    "for col in paf_df.columns:\n",
    "    if 'max_total_qidentity' in col:\n",
    "        max_total_qident_cols.append(col)\n",
    "\n",
    "paf_df[max_total_qident_cols] = paf_df[max_total_qident_cols].fillna(0)\n",
    "paf_df['max_total_qidentity'] = paf_df[max_total_qident_cols].max(axis=1)\n",
    "\n",
    "def high_low_identity(row):\n",
    "    if row['max_total_qidentity'] > 0.90:\n",
    "        return('HI')\n",
    "    elif row['max_total_qidentity'] > 0.70 and row['max_total_qidentity'] <= 0.90 :\n",
    "        return('MI')\n",
    "    else:\n",
    "        return('LI')\n",
    "    \n",
    "paf_df['ilevel'] = paf_df.apply(high_low_identity, axis=1)\n",
    "paf_df['org_map'] = paf_df[max_total_qident_cols].idxmax(axis=1).str.replace('.max_total_qidentity', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#paf_df.sort_values('TGVEG.mito.max_aln_qidentity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>qlen</th>\n",
       "      <th>TGVEG.mito.max_aln_qidentity</th>\n",
       "      <th>TGVEG.mito.max_coverage</th>\n",
       "      <th>TGVEG.mito.min_divergence</th>\n",
       "      <th>TGVEG.mito.max_total_qidentity</th>\n",
       "      <th>TGA4.mito.max_aln_qidentity</th>\n",
       "      <th>TGA4.mito.max_coverage</th>\n",
       "      <th>TGA4.mito.min_divergence</th>\n",
       "      <th>TGA4.mito.max_total_qidentity</th>\n",
       "      <th>...</th>\n",
       "      <th>TgRH_pasteur.mito.max_coverage</th>\n",
       "      <th>TgRH_pasteur.mito.min_divergence</th>\n",
       "      <th>TgRH_pasteur.mito.max_total_qidentity</th>\n",
       "      <th>Ncanium_LIV.mito.max_aln_qidentity</th>\n",
       "      <th>Ncanium_LIV.mito.max_coverage</th>\n",
       "      <th>Ncanium_LIV.mito.min_divergence</th>\n",
       "      <th>Ncanium_LIV.mito.max_total_qidentity</th>\n",
       "      <th>max_total_qidentity</th>\n",
       "      <th>ilevel</th>\n",
       "      <th>org_map</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Tg_mtDNA_V</td>\n",
       "      <td>161</td>\n",
       "      <td>0.475177</td>\n",
       "      <td>0.875776</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.475177</td>\n",
       "      <td>0.581560</td>\n",
       "      <td>0.875776</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.581560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.875776</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>0.723404</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.723404</td>\n",
       "      <td>MI</td>\n",
       "      <td>TgRH_pasteur.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tg_mtDNA_E</td>\n",
       "      <td>683</td>\n",
       "      <td>0.598616</td>\n",
       "      <td>0.846266</td>\n",
       "      <td>0.0664</td>\n",
       "      <td>0.598616</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995608</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.995608</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.394366</td>\n",
       "      <td>0.935578</td>\n",
       "      <td>0.0947</td>\n",
       "      <td>0.394366</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tg_mtDNA_F</td>\n",
       "      <td>179</td>\n",
       "      <td>0.609272</td>\n",
       "      <td>0.843575</td>\n",
       "      <td>0.0749</td>\n",
       "      <td>0.609272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.972067</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.972067</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.747899</td>\n",
       "      <td>0.709497</td>\n",
       "      <td>0.0141</td>\n",
       "      <td>0.747899</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Tg_mtDNA_K</td>\n",
       "      <td>445</td>\n",
       "      <td>0.677725</td>\n",
       "      <td>0.474157</td>\n",
       "      <td>0.0588</td>\n",
       "      <td>0.677725</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995506</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.995506</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801242</td>\n",
       "      <td>0.462921</td>\n",
       "      <td>0.0323</td>\n",
       "      <td>0.801242</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tg_mtDNA_I</td>\n",
       "      <td>204</td>\n",
       "      <td>0.862745</td>\n",
       "      <td>0.534314</td>\n",
       "      <td>0.0376</td>\n",
       "      <td>0.862745</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.980392</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980392</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.406863</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Tg_mtDNA_T</td>\n",
       "      <td>279</td>\n",
       "      <td>0.907895</td>\n",
       "      <td>0.706093</td>\n",
       "      <td>0.0297</td>\n",
       "      <td>0.907895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985663</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.985663</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.790210</td>\n",
       "      <td>0.512545</td>\n",
       "      <td>0.0294</td>\n",
       "      <td>0.790210</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tg_mtDNA_C</td>\n",
       "      <td>1050</td>\n",
       "      <td>0.930233</td>\n",
       "      <td>0.301905</td>\n",
       "      <td>0.0267</td>\n",
       "      <td>0.930233</td>\n",
       "      <td>0.982456</td>\n",
       "      <td>0.977143</td>\n",
       "      <td>0.0031</td>\n",
       "      <td>0.982456</td>\n",
       "      <td>...</td>\n",
       "      <td>0.996190</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>0.999044</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.240952</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>Ncanium_LIV.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Tg_mtDNA_S</td>\n",
       "      <td>205</td>\n",
       "      <td>0.932292</td>\n",
       "      <td>0.936585</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.932292</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.975610</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.975610</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.947917</td>\n",
       "      <td>0.468293</td>\n",
       "      <td>0.0085</td>\n",
       "      <td>0.947917</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Tg_mtDNA_U</td>\n",
       "      <td>354</td>\n",
       "      <td>0.971698</td>\n",
       "      <td>0.703390</td>\n",
       "      <td>0.0127</td>\n",
       "      <td>0.971698</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.994350</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.994350</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.585799</td>\n",
       "      <td>0.954802</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.585799</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Tg_mtDNA_P</td>\n",
       "      <td>184</td>\n",
       "      <td>0.972222</td>\n",
       "      <td>0.586957</td>\n",
       "      <td>0.0148</td>\n",
       "      <td>0.972222</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.961957</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.961957</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.895833</td>\n",
       "      <td>0.586957</td>\n",
       "      <td>0.0499</td>\n",
       "      <td>0.895833</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Tg_mtDNA_Q</td>\n",
       "      <td>65</td>\n",
       "      <td>0.982456</td>\n",
       "      <td>0.876923</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.982456</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.938462</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.938462</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.786885</td>\n",
       "      <td>0.938462</td>\n",
       "      <td>0.0148</td>\n",
       "      <td>0.786885</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Tg_mtDNA_O</td>\n",
       "      <td>86</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.686047</td>\n",
       "      <td>0.0089</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.918605</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.918605</td>\n",
       "      <td>0.0028</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.872093</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Tg_mtDNA_N</td>\n",
       "      <td>166</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.614458</td>\n",
       "      <td>0.0021</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981928</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981928</td>\n",
       "      <td>0.0012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.815534</td>\n",
       "      <td>0.716867</td>\n",
       "      <td>0.0434</td>\n",
       "      <td>0.815534</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tg_mtDNA_A</td>\n",
       "      <td>196</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.709184</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.578378</td>\n",
       "      <td>0.943878</td>\n",
       "      <td>0.0530</td>\n",
       "      <td>0.578378</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Tg_mtDNA_M</td>\n",
       "      <td>754</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.574271</td>\n",
       "      <td>0.0031</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.992042</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.992042</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.408602</td>\n",
       "      <td>0.986737</td>\n",
       "      <td>0.0785</td>\n",
       "      <td>0.408602</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Tg_mtDNA_L</td>\n",
       "      <td>159</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.943396</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.974843</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.974843</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.419355</td>\n",
       "      <td>0.779874</td>\n",
       "      <td>0.0617</td>\n",
       "      <td>0.419355</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tg_mtDNA_J</td>\n",
       "      <td>85</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.588235</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tg_mtDNA_H</td>\n",
       "      <td>447</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.407159</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.975057</td>\n",
       "      <td>0.986577</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0.975057</td>\n",
       "      <td>...</td>\n",
       "      <td>0.986577</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.993197</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.912752</td>\n",
       "      <td>0.0028</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tg_mtDNA_Fp</td>\n",
       "      <td>91</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.780220</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945055</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.945055</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.560976</td>\n",
       "      <td>0.901099</td>\n",
       "      <td>0.0297</td>\n",
       "      <td>0.560976</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tg_mtDNA_D</td>\n",
       "      <td>82</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.841463</td>\n",
       "      <td>0.0089</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963415</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.963415</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.743902</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Tg_mtDNA_Mp</td>\n",
       "      <td>104</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.0060</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.942308</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.942308</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.617021</td>\n",
       "      <td>0.903846</td>\n",
       "      <td>0.0536</td>\n",
       "      <td>0.617021</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Tg_mtDNA_Kp</td>\n",
       "      <td>112</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.732143</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.851852</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.0173</td>\n",
       "      <td>0.851852</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tg_mtDNA_B</td>\n",
       "      <td>40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Tg_mtDNA_R</td>\n",
       "      <td>85</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.695652</td>\n",
       "      <td>0.811765</td>\n",
       "      <td>0.0267</td>\n",
       "      <td>0.695652</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          qname  qlen  TGVEG.mito.max_aln_qidentity  TGVEG.mito.max_coverage  \\\n",
       "23   Tg_mtDNA_V   161                      0.475177                 0.875776   \n",
       "4    Tg_mtDNA_E   683                      0.598616                 0.846266   \n",
       "5    Tg_mtDNA_F   179                      0.609272                 0.843575   \n",
       "10   Tg_mtDNA_K   445                      0.677725                 0.474157   \n",
       "8    Tg_mtDNA_I   204                      0.862745                 0.534314   \n",
       "21   Tg_mtDNA_T   279                      0.907895                 0.706093   \n",
       "2    Tg_mtDNA_C  1050                      0.930233                 0.301905   \n",
       "20   Tg_mtDNA_S   205                      0.932292                 0.936585   \n",
       "22   Tg_mtDNA_U   354                      0.971698                 0.703390   \n",
       "17   Tg_mtDNA_P   184                      0.972222                 0.586957   \n",
       "18   Tg_mtDNA_Q    65                      0.982456                 0.876923   \n",
       "16   Tg_mtDNA_O    86                      1.000000                 0.686047   \n",
       "15   Tg_mtDNA_N   166                      1.000000                 0.614458   \n",
       "0    Tg_mtDNA_A   196                      1.000000                 0.709184   \n",
       "13   Tg_mtDNA_M   754                      1.000000                 0.574271   \n",
       "12   Tg_mtDNA_L   159                      1.000000                 0.943396   \n",
       "9    Tg_mtDNA_J    85                      1.000000                 0.588235   \n",
       "7    Tg_mtDNA_H   447                      1.000000                 0.407159   \n",
       "6   Tg_mtDNA_Fp    91                      1.000000                 0.780220   \n",
       "3    Tg_mtDNA_D    82                      1.000000                 0.841463   \n",
       "14  Tg_mtDNA_Mp   104                      1.000000                 0.769231   \n",
       "11  Tg_mtDNA_Kp   112                      1.000000                 0.732143   \n",
       "1    Tg_mtDNA_B    40                           NaN                      NaN   \n",
       "19   Tg_mtDNA_R    85                           NaN                      NaN   \n",
       "\n",
       "    TGVEG.mito.min_divergence  TGVEG.mito.max_total_qidentity  \\\n",
       "23                     0.0049                        0.475177   \n",
       "4                      0.0664                        0.598616   \n",
       "5                      0.0749                        0.609272   \n",
       "10                     0.0588                        0.677725   \n",
       "8                      0.0376                        0.862745   \n",
       "21                     0.0297                        0.907895   \n",
       "2                      0.0267                        0.930233   \n",
       "20                     0.0251                        0.932292   \n",
       "22                     0.0127                        0.971698   \n",
       "17                     0.0148                        0.972222   \n",
       "18                     0.0308                        0.982456   \n",
       "16                     0.0089                        1.000000   \n",
       "15                     0.0021                        1.000000   \n",
       "0                      0.0000                        1.000000   \n",
       "13                     0.0031                        1.000000   \n",
       "12                     0.0016                        1.000000   \n",
       "9                      0.0078                        1.000000   \n",
       "7                      0.0095                        1.000000   \n",
       "6                      0.0036                        1.000000   \n",
       "3                      0.0089                        1.000000   \n",
       "14                     0.0060                        1.000000   \n",
       "11                     0.0026                        1.000000   \n",
       "1                         NaN                        0.000000   \n",
       "19                        NaN                        0.000000   \n",
       "\n",
       "    TGA4.mito.max_aln_qidentity  TGA4.mito.max_coverage  \\\n",
       "23                     0.581560                0.875776   \n",
       "4                      1.000000                0.995608   \n",
       "5                      1.000000                0.972067   \n",
       "10                     1.000000                0.995506   \n",
       "8                      1.000000                0.980392   \n",
       "21                     1.000000                0.985663   \n",
       "2                      0.982456                0.977143   \n",
       "20                     1.000000                0.975610   \n",
       "22                     1.000000                0.994350   \n",
       "17                     1.000000                0.961957   \n",
       "18                     1.000000                0.938462   \n",
       "16                     1.000000                0.918605   \n",
       "15                     1.000000                0.981928   \n",
       "0                      1.000000                1.000000   \n",
       "13                     1.000000                0.992042   \n",
       "12                     1.000000                0.974843   \n",
       "9                           NaN                     NaN   \n",
       "7                      0.975057                0.986577   \n",
       "6                      1.000000                0.945055   \n",
       "3                      1.000000                0.963415   \n",
       "14                     1.000000                0.942308   \n",
       "11                     1.000000                0.964286   \n",
       "1                           NaN                     NaN   \n",
       "19                     1.000000                0.941176   \n",
       "\n",
       "    TGA4.mito.min_divergence  TGA4.mito.max_total_qidentity  ...  \\\n",
       "23                    0.0040                       0.581560  ...   \n",
       "4                     0.0003                       1.000000  ...   \n",
       "5                     0.0013                       1.000000  ...   \n",
       "10                    0.0004                       1.000000  ...   \n",
       "8                     0.0009                       1.000000  ...   \n",
       "21                    0.0008                       1.000000  ...   \n",
       "2                     0.0031                       0.982456  ...   \n",
       "20                    0.0010                       1.000000  ...   \n",
       "22                    0.0006                       1.000000  ...   \n",
       "17                    0.0012                       1.000000  ...   \n",
       "18                    0.0040                       1.000000  ...   \n",
       "16                    0.0027                       1.000000  ...   \n",
       "15                    0.0012                       1.000000  ...   \n",
       "0                     0.0010                       1.000000  ...   \n",
       "13                    0.0003                       1.000000  ...   \n",
       "12                    0.0015                       1.000000  ...   \n",
       "9                        NaN                       0.000000  ...   \n",
       "7                     0.0033                       0.975057  ...   \n",
       "6                     0.0027                       1.000000  ...   \n",
       "3                     0.0030                       1.000000  ...   \n",
       "14                    0.0025                       1.000000  ...   \n",
       "11                    0.0020                       1.000000  ...   \n",
       "1                        NaN                       0.000000  ...   \n",
       "19                    0.0030                       1.000000  ...   \n",
       "\n",
       "    TgRH_pasteur.mito.max_coverage  TgRH_pasteur.mito.min_divergence  \\\n",
       "23                        0.875776                            0.0036   \n",
       "4                         0.995608                            0.0003   \n",
       "5                         0.972067                            0.0013   \n",
       "10                        0.995506                            0.0004   \n",
       "8                         0.980392                            0.0009   \n",
       "21                        0.985663                            0.0008   \n",
       "2                         0.996190                            0.0007   \n",
       "20                        0.975610                            0.0010   \n",
       "22                        0.994350                            0.0006   \n",
       "17                        0.961957                            0.0012   \n",
       "18                        0.938462                            0.0040   \n",
       "16                        0.918605                            0.0028   \n",
       "15                        0.981928                            0.0012   \n",
       "0                         1.000000                            0.0010   \n",
       "13                        0.992042                            0.0003   \n",
       "12                        0.974843                            0.0015   \n",
       "9                              NaN                               NaN   \n",
       "7                         0.986577                            0.0005   \n",
       "6                         0.945055                            0.0027   \n",
       "3                         0.963415                            0.0030   \n",
       "14                        0.942308                            0.0025   \n",
       "11                        0.964286                            0.0020   \n",
       "1                              NaN                               NaN   \n",
       "19                        0.941176                            0.0030   \n",
       "\n",
       "    TgRH_pasteur.mito.max_total_qidentity  Ncanium_LIV.mito.max_aln_qidentity  \\\n",
       "23                               0.723404                                 NaN   \n",
       "4                                1.000000                            0.394366   \n",
       "5                                1.000000                            0.747899   \n",
       "10                               1.000000                            0.801242   \n",
       "8                                1.000000                            1.000000   \n",
       "21                               1.000000                            0.790210   \n",
       "2                                0.999044                            1.000000   \n",
       "20                               1.000000                            0.947917   \n",
       "22                               1.000000                            0.585799   \n",
       "17                               1.000000                            0.895833   \n",
       "18                               1.000000                            0.786885   \n",
       "16                               1.000000                            1.000000   \n",
       "15                               1.000000                            0.815534   \n",
       "0                                1.000000                            0.578378   \n",
       "13                               1.000000                            0.408602   \n",
       "12                               1.000000                            0.419355   \n",
       "9                                0.000000                                 NaN   \n",
       "7                                0.993197                            1.000000   \n",
       "6                                1.000000                            0.560976   \n",
       "3                                1.000000                            1.000000   \n",
       "14                               1.000000                            0.617021   \n",
       "11                               1.000000                            0.851852   \n",
       "1                                0.000000                                 NaN   \n",
       "19                               1.000000                            0.695652   \n",
       "\n",
       "    Ncanium_LIV.mito.max_coverage  Ncanium_LIV.mito.min_divergence  \\\n",
       "23                            NaN                              NaN   \n",
       "4                        0.935578                           0.0947   \n",
       "5                        0.709497                           0.0141   \n",
       "10                       0.462921                           0.0323   \n",
       "8                        0.406863                           0.0036   \n",
       "21                       0.512545                           0.0294   \n",
       "2                        0.240952                           0.0095   \n",
       "20                       0.468293                           0.0085   \n",
       "22                       0.954802                           0.0737   \n",
       "17                       0.586957                           0.0499   \n",
       "18                       0.938462                           0.0148   \n",
       "16                       0.872093                           0.0040   \n",
       "15                       0.716867                           0.0434   \n",
       "0                        0.943878                           0.0530   \n",
       "13                       0.986737                           0.0785   \n",
       "12                       0.779874                           0.0617   \n",
       "9                             NaN                              NaN   \n",
       "7                        0.912752                           0.0028   \n",
       "6                        0.901099                           0.0297   \n",
       "3                        0.743902                           0.0046   \n",
       "14                       0.903846                           0.0536   \n",
       "11                       0.964286                           0.0173   \n",
       "1                             NaN                              NaN   \n",
       "19                       0.811765                           0.0267   \n",
       "\n",
       "    Ncanium_LIV.mito.max_total_qidentity  max_total_qidentity  ilevel  \\\n",
       "23                              0.000000             0.723404      MI   \n",
       "4                               0.394366             1.000000      HI   \n",
       "5                               0.747899             1.000000      HI   \n",
       "10                              0.801242             1.000000      HI   \n",
       "8                               1.000000             1.000000      HI   \n",
       "21                              0.790210             1.000000      HI   \n",
       "2                               1.000000             1.000000      HI   \n",
       "20                              0.947917             1.000000      HI   \n",
       "22                              0.585799             1.000000      HI   \n",
       "17                              0.895833             1.000000      HI   \n",
       "18                              0.786885             1.000000      HI   \n",
       "16                              1.000000             1.000000      HI   \n",
       "15                              0.815534             1.000000      HI   \n",
       "0                               0.578378             1.000000      HI   \n",
       "13                              0.408602             1.000000      HI   \n",
       "12                              0.419355             1.000000      HI   \n",
       "9                               0.000000             1.000000      HI   \n",
       "7                               1.000000             1.000000      HI   \n",
       "6                               0.560976             1.000000      HI   \n",
       "3                               1.000000             1.000000      HI   \n",
       "14                              0.617021             1.000000      HI   \n",
       "11                              0.851852             1.000000      HI   \n",
       "1                               0.000000             0.000000      LI   \n",
       "19                              0.695652             1.000000      HI   \n",
       "\n",
       "              org_map  \n",
       "23  TgRH_pasteur.mito  \n",
       "4           TGA4.mito  \n",
       "5           TGA4.mito  \n",
       "10          TGA4.mito  \n",
       "8           TGA4.mito  \n",
       "21          TGA4.mito  \n",
       "2    Ncanium_LIV.mito  \n",
       "20          TGA4.mito  \n",
       "22          TGA4.mito  \n",
       "17          TGA4.mito  \n",
       "18          TGA4.mito  \n",
       "16         TGVEG.mito  \n",
       "15         TGVEG.mito  \n",
       "0          TGVEG.mito  \n",
       "13         TGVEG.mito  \n",
       "12         TGVEG.mito  \n",
       "9          TGVEG.mito  \n",
       "7          TGVEG.mito  \n",
       "6          TGVEG.mito  \n",
       "3          TGVEG.mito  \n",
       "14         TGVEG.mito  \n",
       "11         TGVEG.mito  \n",
       "1          TGVEG.mito  \n",
       "19          TGA4.mito  \n",
       "\n",
       "[24 rows x 21 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paf_df.sort_values('TGVEG.mito.max_aln_qidentity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of nucleotides mapped with high (>90%), medium (70-90%) and Low (<70%) identity\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ilevel\n",
       "HI    0.967664\n",
       "LI    0.006435\n",
       "MI    0.025901\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nucleotides in sequences with high (>90%), medium (70-90%) and Low (<70%) identity\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ilevel\n",
       "HI    6015\n",
       "LI      40\n",
       "MI     161\n",
       "Name: qlen, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median size of sequences mapped with high (>90%), medium (70-90%) and Low (<70%) identity\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ilevel\n",
       "HI    181.5\n",
       "LI     40.0\n",
       "MI    161.0\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences with high (>90%), medium (70-90%) and Low (<70%) identity\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ilevel\n",
       "HI    22\n",
       "LI     1\n",
       "MI     1\n",
       "Name: qlen, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('Fraction of nucleotides mapped with high (>90%), medium (70-90%) and Low (<70%) identity')\n",
    "display(paf_df.groupby('ilevel')['qlen'].sum()/total_len)\n",
    "\n",
    "print('Number of nucleotides in sequences with high (>90%), medium (70-90%) and Low (<70%) identity')\n",
    "display(paf_df.groupby('ilevel')['qlen'].sum())\n",
    "\n",
    "print('Median size of sequences mapped with high (>90%), medium (70-90%) and Low (<70%) identity')\n",
    "display(paf_df.groupby('ilevel')['qlen'].median())\n",
    "\n",
    "print('Number of sequences with high (>90%), medium (70-90%) and Low (<70%) identity')\n",
    "display(paf_df.groupby('ilevel')['qlen'].count())\n",
    "\n",
    "\n",
    "hi_df = paf_df[paf_df['ilevel'] == 'HI']\n",
    "mi_df = paf_df[paf_df['ilevel'] == 'MI']\n",
    "li_df = paf_df[paf_df['ilevel'] == 'LI']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fraction of nucleotides mapped with high (>90%), medium (70-90%) and Low (<70%) identity\n",
    "ilevel\n",
    "HI    0.973684\n",
    "LI    0.012504\n",
    "MI    0.013812\n",
    "Name: qlen, dtype: float64\n",
    "Number of nucleotides in sequences with high (>90%), medium (70-90%) and Low (<70%) identity\n",
    "ilevel\n",
    "HI    1557456813\n",
    "LI      20000564\n",
    "MI      22092434\n",
    "Name: qlen, dtype: int64\n",
    "Median size of sequences mapped with high (>90%), medium (70-90%) and Low (<70%) identity\n",
    "ilevel\n",
    "HI    9292.0\n",
    "LI    1372.0\n",
    "MI    6342.0\n",
    "Name: qlen, dtype: float64\n",
    "Number of sequences with high (>90%), medium (70-90%) and Low (<70%) identity\n",
    "ilevel\n",
    "HI    121931\n",
    "LI      4843\n",
    "MI      2102\n",
    "Name: qlen, dtype: int64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reads mapped with high identity (>90%)\n",
      "Fraction of mapped nucleotides asigned to each reference\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "org_map\n",
       "Ncanium_LIV.mito    0.174564\n",
       "TGA4.mito           0.446052\n",
       "TGVEG.mito          0.379385\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median sequence size asigned to each reference\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "org_map\n",
       "Ncanium_LIV.mito    1050.0\n",
       "TGA4.mito            204.5\n",
       "TGVEG.mito           112.0\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('Reads mapped with high identity (>90%)')\n",
    "\n",
    "print('Fraction of mapped nucleotides asigned to each reference')\n",
    "display(hi_df.groupby('org_map')['qlen'].sum()/hi_df['qlen'].sum())\n",
    "\n",
    "print('Median sequence size asigned to each reference')\n",
    "display(hi_df.groupby('org_map')['qlen'].median())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reads mapped with high identity (>90%)\n",
    "Fraction of mapped nucleotides asigned to each reference\n",
    "org_map\n",
    "Chlorocebus_sabaeus    0.016857\n",
    "TGA4                   0.032556\n",
    "TGVEG                  0.035012\n",
    "TG_APlast              0.001050\n",
    "TgRH_pasteur           0.914526\n",
    "Name: qlen, dtype: float64\n",
    "Median sequence size asigned to each reference\n",
    "org_map\n",
    "Chlorocebus_sabaeus    16879.0\n",
    "TGA4                    8681.0\n",
    "TGVEG                   6475.0\n",
    "TG_APlast              19686.0\n",
    "TgRH_pasteur            9368.5\n",
    "Name: qlen, dtype: float64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reads mapped with medium identity (70-90%)\n",
      "Fraction of mapped nucleotides asigned to each reference\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "org_map\n",
       "TgRH_pasteur.mito   NaN\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median sequence size asigned to each reference\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "org_map\n",
       "TgRH_pasteur.mito   NaN\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('Reads mapped with medium identity (70-90%)')\n",
    "\n",
    "print('Fraction of mapped nucleotides asigned to each reference')\n",
    "display(mi_df.groupby('org_map')['qlen'].sum()/mi_df['qlen'].sum())\n",
    "\n",
    "print('Median sequence size asigned to each reference')\n",
    "display(mi_df.groupby('org_map')['qlen'].median())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reads mapped with medium identity (70-90%)\n",
    "Fraction of mapped nucleotides asigned to each reference\n",
    "org_map\n",
    "Chlorocebus_sabaeus    0.004433\n",
    "TGA4                   0.215406\n",
    "TGVEG                  0.317919\n",
    "TgRH_pasteur           0.462243\n",
    "Name: qlen, dtype: float64\n",
    "Median sequence size asigned to each reference\n",
    "org_map\n",
    "Chlorocebus_sabaeus    3761.0\n",
    "TGA4                   4688.0\n",
    "TGVEG                  8199.0\n",
    "TgRH_pasteur           7190.0\n",
    "Name: qlen, dtype: float64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reads mapped with low identity (<70%)\n",
      "Fraction of mapped nucleotides asigned to each reference\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "org_map\n",
       "TGVEG.mito    1.0\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median sequence size asigned to each reference\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "org_map\n",
       "TGVEG.mito    8938.5\n",
       "Name: qlen, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('Reads mapped with low identity (<70%)')\n",
    "\n",
    "print('Fraction of mapped nucleotides asigned to each reference')\n",
    "display(li_df.groupby('org_map')['qlen'].sum()/li_df['qlen'].sum())\n",
    "\n",
    "print('Median sequence size asigned to each reference')\n",
    "display(li_df.groupby('org_map')['qlen'].median())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reads mapped with low identity (<70%)\n",
    "Fraction of mapped nucleotides asigned to each reference\n",
    "org_map\n",
    "Chlorocebus_sabaeus    0.566965\n",
    "TGA4                   0.122387\n",
    "TGVEG                  0.278444\n",
    "TgRH_pasteur           0.032203\n",
    "Name: qlen, dtype: float64\n",
    "Median sequence size asigned to each reference\n",
    "org_map\n",
    "Chlorocebus_sabaeus    9180.0\n",
    "TGA4                   7804.0\n",
    "TGVEG                  1268.0\n",
    "TgRH_pasteur           8966.0\n",
    "Name: qlen, dtype: float64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [],
   "source": [
    "li_ids = set(li_df['qname'])\n",
    "with open(output_unmapped_reads_path, 'w') as file:\n",
    "    for record in SeqIO.parse(input_query_fasta_path, \"fasta\"):\n",
    "        if record.id in li_ids:\n",
    "            seq_head=f'>{record.id}\\n'\n",
    "            seq_body=f'{record.seq}\\n'\n",
    "            file.write(seq_head)\n",
    "            file.write(seq_body)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section with test/squetch code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>qlen</th>\n",
       "      <th>TGVEG.mito.max_coverage</th>\n",
       "      <th>TGVEG.mito.min_divergence</th>\n",
       "      <th>TGA4.mito.max_coverage</th>\n",
       "      <th>TGA4.mito.min_divergence</th>\n",
       "      <th>TgRH_pasteur.mito.max_coverage</th>\n",
       "      <th>TgRH_pasteur.mito.min_divergence</th>\n",
       "      <th>Ncanium_LIV.mito.max_coverage</th>\n",
       "      <th>Ncanium_LIV.mito.min_divergence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>51912</td>\n",
       "      <td>108517.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>104371</td>\n",
       "      <td>105404.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22719</td>\n",
       "      <td>99502.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28700</td>\n",
       "      <td>96911.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32082</td>\n",
       "      <td>96252.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>42046</td>\n",
       "      <td>95866.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index      qlen  TGVEG.mito.max_coverage  TGVEG.mito.min_divergence  \\\n",
       "0   51912  108517.0                      NaN                        NaN   \n",
       "1  104371  105404.0                      NaN                        NaN   \n",
       "2   22719   99502.0                      NaN                        NaN   \n",
       "3   28700   96911.0                      NaN                        NaN   \n",
       "4   32082   96252.0                      NaN                        NaN   \n",
       "5   42046   95866.0                      NaN                        NaN   \n",
       "\n",
       "   TGA4.mito.max_coverage  TGA4.mito.min_divergence  \\\n",
       "0                     NaN                       NaN   \n",
       "1                     NaN                       NaN   \n",
       "2                     NaN                       NaN   \n",
       "3                     NaN                       NaN   \n",
       "4                     NaN                       NaN   \n",
       "5                     NaN                       NaN   \n",
       "\n",
       "   TgRH_pasteur.mito.max_coverage  TgRH_pasteur.mito.min_divergence  \\\n",
       "0                             NaN                               NaN   \n",
       "1                             NaN                               NaN   \n",
       "2                             NaN                               NaN   \n",
       "3                             NaN                               NaN   \n",
       "4                             NaN                               NaN   \n",
       "5                             NaN                               NaN   \n",
       "\n",
       "   Ncanium_LIV.mito.max_coverage  Ncanium_LIV.mito.min_divergence  \n",
       "0                            NaN                              NaN  \n",
       "1                            NaN                              NaN  \n",
       "2                            NaN                              NaN  \n",
       "3                            NaN                              NaN  \n",
       "4                            NaN                              NaN  \n",
       "5                            NaN                              NaN  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paf_df_sort = paf_df.filter(regex='qlen|min_divergence|max_coverage').sort_values(by='qlen',ascending=False)\n",
    "paf_df_sort = paf_df_sort.reset_index()\n",
    "paf_df_sort.head(6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>qlen</th>\n",
       "      <th>TGVEG.mito.max_aln_qidentity</th>\n",
       "      <th>TGVEG.mito.max_coverage</th>\n",
       "      <th>TGVEG.mito.min_divergence</th>\n",
       "      <th>TGVEG.mito.max_total_qidentity</th>\n",
       "      <th>TGA4.mito.max_aln_qidentity</th>\n",
       "      <th>TGA4.mito.max_coverage</th>\n",
       "      <th>TGA4.mito.min_divergence</th>\n",
       "      <th>TGA4.mito.max_total_qidentity</th>\n",
       "      <th>...</th>\n",
       "      <th>TgRH_pasteur.mito.max_coverage</th>\n",
       "      <th>TgRH_pasteur.mito.min_divergence</th>\n",
       "      <th>TgRH_pasteur.mito.max_total_qidentity</th>\n",
       "      <th>Ncanium_LIV.mito.max_aln_qidentity</th>\n",
       "      <th>Ncanium_LIV.mito.max_coverage</th>\n",
       "      <th>Ncanium_LIV.mito.min_divergence</th>\n",
       "      <th>Ncanium_LIV.mito.max_total_qidentity</th>\n",
       "      <th>max_total_qidentity</th>\n",
       "      <th>ilevel</th>\n",
       "      <th>org_map</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51912</th>\n",
       "      <td>6743d56d-6b55-42f8-898d-ca6ac80a5031</td>\n",
       "      <td>108517.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104371</th>\n",
       "      <td>cf81ab03-c55c-4fa1-9969-a91722b59b2b</td>\n",
       "      <td>105404.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22719</th>\n",
       "      <td>2d299152-0621-4079-b43b-7da225d6eb95</td>\n",
       "      <td>99502.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28700</th>\n",
       "      <td>3910ae3b-a42d-4096-9e19-9b632217ec63</td>\n",
       "      <td>96911.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32082</th>\n",
       "      <td>3fbcd979-a862-45bd-9283-ca110a54b733</td>\n",
       "      <td>96252.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LI</td>\n",
       "      <td>TGVEG.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80414</th>\n",
       "      <td>Tg_mtDNA_R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.695652</td>\n",
       "      <td>0.811765</td>\n",
       "      <td>0.0267</td>\n",
       "      <td>0.695652</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80415</th>\n",
       "      <td>Tg_mtDNA_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.932292</td>\n",
       "      <td>0.936585</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.932292</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.975610</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.975610</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.947917</td>\n",
       "      <td>0.468293</td>\n",
       "      <td>0.0085</td>\n",
       "      <td>0.947917</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80416</th>\n",
       "      <td>Tg_mtDNA_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.907895</td>\n",
       "      <td>0.706093</td>\n",
       "      <td>0.0297</td>\n",
       "      <td>0.907895</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.985663</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.985663</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.790210</td>\n",
       "      <td>0.512545</td>\n",
       "      <td>0.0294</td>\n",
       "      <td>0.790210</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80417</th>\n",
       "      <td>Tg_mtDNA_U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.971698</td>\n",
       "      <td>0.703390</td>\n",
       "      <td>0.0127</td>\n",
       "      <td>0.971698</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.994350</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.994350</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.585799</td>\n",
       "      <td>0.954802</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.585799</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>HI</td>\n",
       "      <td>TGA4.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80418</th>\n",
       "      <td>Tg_mtDNA_V</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.475177</td>\n",
       "      <td>0.875776</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.475177</td>\n",
       "      <td>0.58156</td>\n",
       "      <td>0.875776</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.58156</td>\n",
       "      <td>...</td>\n",
       "      <td>0.875776</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>0.723404</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.723404</td>\n",
       "      <td>MI</td>\n",
       "      <td>TgRH_pasteur.mito</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128899 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       qname      qlen  \\\n",
       "51912   6743d56d-6b55-42f8-898d-ca6ac80a5031  108517.0   \n",
       "104371  cf81ab03-c55c-4fa1-9969-a91722b59b2b  105404.0   \n",
       "22719   2d299152-0621-4079-b43b-7da225d6eb95   99502.0   \n",
       "28700   3910ae3b-a42d-4096-9e19-9b632217ec63   96911.0   \n",
       "32082   3fbcd979-a862-45bd-9283-ca110a54b733   96252.0   \n",
       "...                                      ...       ...   \n",
       "80414                             Tg_mtDNA_R       NaN   \n",
       "80415                             Tg_mtDNA_S       NaN   \n",
       "80416                             Tg_mtDNA_T       NaN   \n",
       "80417                             Tg_mtDNA_U       NaN   \n",
       "80418                             Tg_mtDNA_V       NaN   \n",
       "\n",
       "        TGVEG.mito.max_aln_qidentity  TGVEG.mito.max_coverage  \\\n",
       "51912                            NaN                      NaN   \n",
       "104371                           NaN                      NaN   \n",
       "22719                            NaN                      NaN   \n",
       "28700                            NaN                      NaN   \n",
       "32082                            NaN                      NaN   \n",
       "...                              ...                      ...   \n",
       "80414                            NaN                      NaN   \n",
       "80415                       0.932292                 0.936585   \n",
       "80416                       0.907895                 0.706093   \n",
       "80417                       0.971698                 0.703390   \n",
       "80418                       0.475177                 0.875776   \n",
       "\n",
       "        TGVEG.mito.min_divergence  TGVEG.mito.max_total_qidentity  \\\n",
       "51912                         NaN                        0.000000   \n",
       "104371                        NaN                        0.000000   \n",
       "22719                         NaN                        0.000000   \n",
       "28700                         NaN                        0.000000   \n",
       "32082                         NaN                        0.000000   \n",
       "...                           ...                             ...   \n",
       "80414                         NaN                        0.000000   \n",
       "80415                      0.0251                        0.932292   \n",
       "80416                      0.0297                        0.907895   \n",
       "80417                      0.0127                        0.971698   \n",
       "80418                      0.0049                        0.475177   \n",
       "\n",
       "        TGA4.mito.max_aln_qidentity  TGA4.mito.max_coverage  \\\n",
       "51912                           NaN                     NaN   \n",
       "104371                          NaN                     NaN   \n",
       "22719                           NaN                     NaN   \n",
       "28700                           NaN                     NaN   \n",
       "32082                           NaN                     NaN   \n",
       "...                             ...                     ...   \n",
       "80414                       1.00000                0.941176   \n",
       "80415                       1.00000                0.975610   \n",
       "80416                       1.00000                0.985663   \n",
       "80417                       1.00000                0.994350   \n",
       "80418                       0.58156                0.875776   \n",
       "\n",
       "        TGA4.mito.min_divergence  TGA4.mito.max_total_qidentity  ...  \\\n",
       "51912                        NaN                        0.00000  ...   \n",
       "104371                       NaN                        0.00000  ...   \n",
       "22719                        NaN                        0.00000  ...   \n",
       "28700                        NaN                        0.00000  ...   \n",
       "32082                        NaN                        0.00000  ...   \n",
       "...                          ...                            ...  ...   \n",
       "80414                     0.0030                        1.00000  ...   \n",
       "80415                     0.0010                        1.00000  ...   \n",
       "80416                     0.0008                        1.00000  ...   \n",
       "80417                     0.0006                        1.00000  ...   \n",
       "80418                     0.0040                        0.58156  ...   \n",
       "\n",
       "        TgRH_pasteur.mito.max_coverage  TgRH_pasteur.mito.min_divergence  \\\n",
       "51912                              NaN                               NaN   \n",
       "104371                             NaN                               NaN   \n",
       "22719                              NaN                               NaN   \n",
       "28700                              NaN                               NaN   \n",
       "32082                              NaN                               NaN   \n",
       "...                                ...                               ...   \n",
       "80414                         0.941176                            0.0030   \n",
       "80415                         0.975610                            0.0010   \n",
       "80416                         0.985663                            0.0008   \n",
       "80417                         0.994350                            0.0006   \n",
       "80418                         0.875776                            0.0036   \n",
       "\n",
       "        TgRH_pasteur.mito.max_total_qidentity  \\\n",
       "51912                                0.000000   \n",
       "104371                               0.000000   \n",
       "22719                                0.000000   \n",
       "28700                                0.000000   \n",
       "32082                                0.000000   \n",
       "...                                       ...   \n",
       "80414                                1.000000   \n",
       "80415                                1.000000   \n",
       "80416                                1.000000   \n",
       "80417                                1.000000   \n",
       "80418                                0.723404   \n",
       "\n",
       "        Ncanium_LIV.mito.max_aln_qidentity  Ncanium_LIV.mito.max_coverage  \\\n",
       "51912                                  NaN                            NaN   \n",
       "104371                                 NaN                            NaN   \n",
       "22719                                  NaN                            NaN   \n",
       "28700                                  NaN                            NaN   \n",
       "32082                                  NaN                            NaN   \n",
       "...                                    ...                            ...   \n",
       "80414                             0.695652                       0.811765   \n",
       "80415                             0.947917                       0.468293   \n",
       "80416                             0.790210                       0.512545   \n",
       "80417                             0.585799                       0.954802   \n",
       "80418                                  NaN                            NaN   \n",
       "\n",
       "        Ncanium_LIV.mito.min_divergence  Ncanium_LIV.mito.max_total_qidentity  \\\n",
       "51912                               NaN                              0.000000   \n",
       "104371                              NaN                              0.000000   \n",
       "22719                               NaN                              0.000000   \n",
       "28700                               NaN                              0.000000   \n",
       "32082                               NaN                              0.000000   \n",
       "...                                 ...                                   ...   \n",
       "80414                            0.0267                              0.695652   \n",
       "80415                            0.0085                              0.947917   \n",
       "80416                            0.0294                              0.790210   \n",
       "80417                            0.0737                              0.585799   \n",
       "80418                               NaN                              0.000000   \n",
       "\n",
       "        max_total_qidentity  ilevel            org_map  \n",
       "51912              0.000000      LI         TGVEG.mito  \n",
       "104371             0.000000      LI         TGVEG.mito  \n",
       "22719              0.000000      LI         TGVEG.mito  \n",
       "28700              0.000000      LI         TGVEG.mito  \n",
       "32082              0.000000      LI         TGVEG.mito  \n",
       "...                     ...     ...                ...  \n",
       "80414              1.000000      HI          TGA4.mito  \n",
       "80415              1.000000      HI          TGA4.mito  \n",
       "80416              1.000000      HI          TGA4.mito  \n",
       "80417              1.000000      HI          TGA4.mito  \n",
       "80418              0.723404      MI  TgRH_pasteur.mito  \n",
       "\n",
       "[128899 rows x 21 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paf_df.sort_values(by='qlen',ascending=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "OE1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
