{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The objective of this script is to make visualisations that shows for each read of the query file the quality of \n",
    "# its alignament between against the different refferences. Also, this script will make barplots that shows the the\n",
    "# quality of the reads mapped in each reference. And plots that shows the quality and quantity of the \n",
    "# reference sequences that were mapped. \n",
    "# \n",
    "# 1 Line plot with mapping qualities of each query sequences against the different references. \n",
    "#   sorted by the qualitiess of the best aligned reference\n",
    "# 2 Barplot showing the percentage of reads mapped in each references, and the quality of the mapped sequences\n",
    "# 3 Barplot showing only for the sequences mapped the quality of its alignament\n",
    "# 4 Try to use different metrics for check quality (Identity, Bitscore)     \n",
    "# 5 Subidive the alignament table, grouping the best read alignamentet to each reference sequencs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jpereira/OEs/OE1/Scripts'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total reads length: 1599549811\n",
      "Sequence number: 128876\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Replace 'your_fasta_file.fasta' with the path to your FASTA file\n",
    "fasta_file = '../toxo_reads/toxo1.trimmedReads.fasta'\n",
    "\n",
    "# Parse the FASTA file and extract the lengths\n",
    "total_len = 0\n",
    "num_records = 0\n",
    "query_lens = {}\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    total_len += len(record.seq)\n",
    "    num_records += 1\n",
    "    query_lens[record.id] = len(record.seq)\n",
    "\n",
    "print(f'Total reads length: {total_len}')\n",
    "print(f'Sequence number: {num_records}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "330996\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_name</th>\n",
       "      <th>query_length</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>relative_strand</th>\n",
       "      <th>target_name</th>\n",
       "      <th>target_length</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "      <th>num_matching_residues</th>\n",
       "      <th>alignment_block_length</th>\n",
       "      <th>mapping_quality</th>\n",
       "      <th>Undf1</th>\n",
       "      <th>Undf2</th>\n",
       "      <th>Undf3</th>\n",
       "      <th>Undf4</th>\n",
       "      <th>Undf5</th>\n",
       "      <th>Undf6</th>\n",
       "      <th>org_name</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>179694</th>\n",
       "      <td>000230c9-5c46-4db0-9fff-06db681304c3</td>\n",
       "      <td>12155</td>\n",
       "      <td>9</td>\n",
       "      <td>12154</td>\n",
       "      <td>+</td>\n",
       "      <td>KI544497.1</td>\n",
       "      <td>4970800</td>\n",
       "      <td>702548</td>\n",
       "      <td>714691</td>\n",
       "      <td>10886</td>\n",
       "      <td>12207</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:1873</td>\n",
       "      <td>s1:i:10861</td>\n",
       "      <td>s2:i:0</td>\n",
       "      <td>dv:f:0.0031</td>\n",
       "      <td>rl:i:153</td>\n",
       "      <td>TGVEG</td>\n",
       "      <td>0.895599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135270</th>\n",
       "      <td>00028b49-baf3-4508-bc07-37dab84feba1</td>\n",
       "      <td>5096</td>\n",
       "      <td>6</td>\n",
       "      <td>5089</td>\n",
       "      <td>+</td>\n",
       "      <td>KI544492.1</td>\n",
       "      <td>6950490</td>\n",
       "      <td>1996854</td>\n",
       "      <td>2001906</td>\n",
       "      <td>4432</td>\n",
       "      <td>5114</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:746</td>\n",
       "      <td>s1:i:4410</td>\n",
       "      <td>s2:i:0</td>\n",
       "      <td>dv:f:0.0034</td>\n",
       "      <td>rl:i:190</td>\n",
       "      <td>TGVEG</td>\n",
       "      <td>0.869702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177654</th>\n",
       "      <td>0003aff6-0ec3-44cc-b1db-0960db1d9132</td>\n",
       "      <td>14828</td>\n",
       "      <td>4</td>\n",
       "      <td>14823</td>\n",
       "      <td>-</td>\n",
       "      <td>KI544497.1</td>\n",
       "      <td>4970800</td>\n",
       "      <td>2401697</td>\n",
       "      <td>2416527</td>\n",
       "      <td>13757</td>\n",
       "      <td>14833</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:2409</td>\n",
       "      <td>s1:i:13752</td>\n",
       "      <td>s2:i:0</td>\n",
       "      <td>dv:f:0.0016</td>\n",
       "      <td>rl:i:15</td>\n",
       "      <td>TGVEG</td>\n",
       "      <td>0.927772</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  query_name  query_length  query_start  \\\n",
       "179694  000230c9-5c46-4db0-9fff-06db681304c3         12155            9   \n",
       "135270  00028b49-baf3-4508-bc07-37dab84feba1          5096            6   \n",
       "177654  0003aff6-0ec3-44cc-b1db-0960db1d9132         14828            4   \n",
       "\n",
       "        query_end relative_strand target_name  target_length  target_start  \\\n",
       "179694      12154               +  KI544497.1        4970800        702548   \n",
       "135270       5089               +  KI544492.1        6950490       1996854   \n",
       "177654      14823               -  KI544497.1        4970800       2401697   \n",
       "\n",
       "        target_end  num_matching_residues  alignment_block_length  \\\n",
       "179694      714691                  10886                   12207   \n",
       "135270     2001906                   4432                    5114   \n",
       "177654     2416527                  13757                   14833   \n",
       "\n",
       "        mapping_quality   Undf1      Undf2       Undf3   Undf4        Undf5  \\\n",
       "179694               60  tp:A:P  cm:i:1873  s1:i:10861  s2:i:0  dv:f:0.0031   \n",
       "135270               60  tp:A:P   cm:i:746   s1:i:4410  s2:i:0  dv:f:0.0034   \n",
       "177654               60  tp:A:P  cm:i:2409  s1:i:13752  s2:i:0  dv:f:0.0016   \n",
       "\n",
       "           Undf6 org_name  identity  \n",
       "179694  rl:i:153    TGVEG  0.895599  \n",
       "135270  rl:i:190    TGVEG  0.869702  \n",
       "177654   rl:i:15    TGVEG  0.927772  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "# Define the path to your PAF file\n",
    "paf_file1_path = '../Data/map_contamination/TGVEG.paf'\n",
    "paf_file2_path = '../Data/map_contamination/TGA4.paf'\n",
    "paf_file3_path = '../Data/map_contamination/TgRH_pasteur.paf'\n",
    "paf_file4_path = '../Data/map_contamination/Chlorocebus_sabaeus.paf'\n",
    "\n",
    "paf_files_path_list = [paf_file1_path, paf_file2_path, paf_file3_path, paf_file4_path]\n",
    "\n",
    "# Define column names based on the PAF specification\n",
    "column_names = [\n",
    "    'query_name', 'query_length', 'query_start', 'query_end',\n",
    "    'relative_strand', 'target_name', 'target_length', 'target_start',\n",
    "    'target_end', 'num_matching_residues', 'alignment_block_length', 'mapping_quality'\n",
    ", 'Undf1','Undf2','Undf3', 'Undf4', 'Undf5', 'Undf6']\n",
    "\n",
    "# Load the PAF file into a DataFrame\n",
    "paf_df = pd.DataFrame()\n",
    "for path in paf_files_path_list:\n",
    "\n",
    "    df = pd.read_csv(path, sep='\\t', header=None, names=column_names)\n",
    "    file_name = os.path.basename(path)\n",
    "    org_name = re.sub('.[^.]*$', '', file_name)\n",
    "    \n",
    "    df['org_name'] = org_name\n",
    "\n",
    "    select_quality = df['mapping_quality'] > 45\n",
    "    select_cover = (df['query_end'] - df['query_start']) / df['query_length'] > 0.9\n",
    "\n",
    "    \n",
    "\n",
    "    select = select_cover & select_quality\n",
    "    df = df[select]\n",
    "\n",
    "    df['identity'] = df['num_matching_residues'] / df['query_length']\n",
    "\n",
    "    # If a query has more than two alignaments (for the same reference),\n",
    "    # select only the one with the greatest identity\n",
    "    max_query_ident = df.groupby('query_name')['identity'].idxmax() \n",
    "    df = df.loc[max_query_ident.values]\n",
    "\n",
    "    paf_df = pd.concat([paf_df, df], axis=0)\n",
    "\n",
    "max_query_ident\n",
    "\n",
    "print(paf_df.shape[0])\n",
    "paf_df.head(3)\n",
    "\n",
    "\n",
    "# Display the first few rows of the DataFrame\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_name</th>\n",
       "      <th>query_length</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>relative_strand</th>\n",
       "      <th>target_name</th>\n",
       "      <th>target_length</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "      <th>num_matching_residues</th>\n",
       "      <th>...</th>\n",
       "      <th>mapping_quality</th>\n",
       "      <th>Undf1</th>\n",
       "      <th>Undf2</th>\n",
       "      <th>Undf3</th>\n",
       "      <th>Undf4</th>\n",
       "      <th>Undf5</th>\n",
       "      <th>Undf6</th>\n",
       "      <th>org_name</th>\n",
       "      <th>identity</th>\n",
       "      <th>max_iden_org_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>000230c9-5c46-4db0-9fff-06db681304c3</td>\n",
       "      <td>12155</td>\n",
       "      <td>9</td>\n",
       "      <td>12154</td>\n",
       "      <td>+</td>\n",
       "      <td>KI544497.1</td>\n",
       "      <td>4970800</td>\n",
       "      <td>702548</td>\n",
       "      <td>714691</td>\n",
       "      <td>10886</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:1873</td>\n",
       "      <td>s1:i:10861</td>\n",
       "      <td>s2:i:0</td>\n",
       "      <td>dv:f:0.0031</td>\n",
       "      <td>rl:i:153</td>\n",
       "      <td>TGVEG</td>\n",
       "      <td>0.895599</td>\n",
       "      <td>TgRH_pasteur</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108515</th>\n",
       "      <td>000230c9-5c46-4db0-9fff-06db681304c3</td>\n",
       "      <td>12155</td>\n",
       "      <td>9</td>\n",
       "      <td>12154</td>\n",
       "      <td>-</td>\n",
       "      <td>CM002047.1</td>\n",
       "      <td>7094428</td>\n",
       "      <td>5964721</td>\n",
       "      <td>5976866</td>\n",
       "      <td>10904</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:1879</td>\n",
       "      <td>s1:i:10880</td>\n",
       "      <td>s2:i:0</td>\n",
       "      <td>dv:f:0.0031</td>\n",
       "      <td>rl:i:83</td>\n",
       "      <td>TGA4</td>\n",
       "      <td>0.897079</td>\n",
       "      <td>TgRH_pasteur</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218614</th>\n",
       "      <td>000230c9-5c46-4db0-9fff-06db681304c3</td>\n",
       "      <td>12155</td>\n",
       "      <td>9</td>\n",
       "      <td>12154</td>\n",
       "      <td>-</td>\n",
       "      <td>CM065588.1</td>\n",
       "      <td>7235803</td>\n",
       "      <td>6100174</td>\n",
       "      <td>6112342</td>\n",
       "      <td>12069</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>tp:A:P</td>\n",
       "      <td>cm:i:2255</td>\n",
       "      <td>s1:i:12064</td>\n",
       "      <td>s2:i:0</td>\n",
       "      <td>dv:f:0.0002</td>\n",
       "      <td>rl:i:113</td>\n",
       "      <td>TgRH_pasteur</td>\n",
       "      <td>0.992925</td>\n",
       "      <td>TgRH_pasteur</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  query_name  query_length  query_start  \\\n",
       "0       000230c9-5c46-4db0-9fff-06db681304c3         12155            9   \n",
       "108515  000230c9-5c46-4db0-9fff-06db681304c3         12155            9   \n",
       "218614  000230c9-5c46-4db0-9fff-06db681304c3         12155            9   \n",
       "\n",
       "        query_end relative_strand target_name  target_length  target_start  \\\n",
       "0           12154               +  KI544497.1        4970800        702548   \n",
       "108515      12154               -  CM002047.1        7094428       5964721   \n",
       "218614      12154               -  CM065588.1        7235803       6100174   \n",
       "\n",
       "        target_end  num_matching_residues  ...  mapping_quality   Undf1  \\\n",
       "0           714691                  10886  ...               60  tp:A:P   \n",
       "108515     5976866                  10904  ...               60  tp:A:P   \n",
       "218614     6112342                  12069  ...               60  tp:A:P   \n",
       "\n",
       "            Undf2       Undf3   Undf4        Undf5     Undf6      org_name  \\\n",
       "0       cm:i:1873  s1:i:10861  s2:i:0  dv:f:0.0031  rl:i:153         TGVEG   \n",
       "108515  cm:i:1879  s1:i:10880  s2:i:0  dv:f:0.0031   rl:i:83          TGA4   \n",
       "218614  cm:i:2255  s1:i:12064  s2:i:0  dv:f:0.0002  rl:i:113  TgRH_pasteur   \n",
       "\n",
       "        identity  max_iden_org_name  \n",
       "0       0.895599       TgRH_pasteur  \n",
       "108515  0.897079       TgRH_pasteur  \n",
       "218614  0.992925       TgRH_pasteur  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paf_df.index = paf_df.reset_index().index\n",
    "id_max_ident = paf_df.groupby('query_name')['identity'].idxmax()\n",
    "\n",
    "query_max_iden_org_df = paf_df.loc[id_max_ident.values, ['query_name','org_name']]\n",
    "query_max_iden_org_df = query_max_iden_org_df.rename(columns={ 'org_name' : 'max_iden_org_name'})\n",
    "\n",
    "paf_max_iden_df = pd.merge(paf_df, query_max_iden_org_df, on='query_name', how='left')\n",
    "paf_max_iden_df.sort_values('query_name').head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "paf_max_iden_df['identity'] = paf_max_iden_df['identity'].round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>query_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>org_name</th>\n",
       "      <th>identity</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Chlorocebus_sabaeus</th>\n",
       "      <th>0.13</th>\n",
       "      <td>7408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.14</th>\n",
       "      <td>4374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.15</th>\n",
       "      <td>3829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.24</th>\n",
       "      <td>1802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.25</th>\n",
       "      <td>1440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">TgRH_pasteur</th>\n",
       "      <th>0.96</th>\n",
       "      <td>39332855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.97</th>\n",
       "      <td>94139272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.98</th>\n",
       "      <td>248560854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.99</th>\n",
       "      <td>676733403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.00</th>\n",
       "      <td>365502443</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              query_length\n",
       "org_name            identity              \n",
       "Chlorocebus_sabaeus 0.13              7408\n",
       "                    0.14              4374\n",
       "                    0.15              3829\n",
       "                    0.24              1802\n",
       "                    0.25              1440\n",
       "...                                    ...\n",
       "TgRH_pasteur        0.96          39332855\n",
       "                    0.97          94139272\n",
       "                    0.98         248560854\n",
       "                    0.99         676733403\n",
       "                    1.00         365502443\n",
       "\n",
       "[288 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_org_idn_seq_len = paf_max_iden_df.groupby(['org_name', 'identity'])['query_length'].sum()\n",
    "len_id_org_df = pd.DataFrame(group_org_idn_seq_len)\n",
    "select_p = len_id_org_df/total_len > 0.05/100\n",
    "\n",
    "#len_id_org_df = len_id_org_df[select_p].dropna()\n",
    "\n",
    "#len_id_org_df.index.get_level_values('org_name').unique()\n",
    "len_id_org_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make bins of 0.1% the total length of sequences\n",
    "# 1: Add to each 'bin' query_length values until is full.\n",
    "# 2: Calculate the identity average of each mean, with it correspondeing quartiles\n",
    "#    -Divide all query lengths by the smallest query_length, round the numbers to 3 decimal places\n",
    "#    -Multiply each query_length by its corresponding identity.\n",
    "#    -Divide all by the bin ocupied size dividev by the smallest query_length\n",
    "# 3: Move to the next bin\n",
    "# 4: Add what is left of the last value of the previus bin to the current bin\n",
    "# 5: Repeat\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chlorocebus_sabaeus', 'TGA4', 'TGVEG', 'TgRH_pasteur'], dtype='object', name='org_name')\n"
     ]
    }
   ],
   "source": [
    "# Extract unique 'org_name' values\n",
    "unique_org_names = group_org_idn_seq_len.index.get_level_values('org_name').unique()\n",
    "\n",
    "print(unique_org_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Organization: Chlorocebus_sabaeus\n",
      "identity\n",
      "0.13    4.631303e-06\n",
      "0.14    2.734519e-06\n",
      "0.15    2.393799e-06\n",
      "0.24    1.126567e-06\n",
      "0.25    9.002533e-07\n",
      "0.29    7.383327e-07\n",
      "0.30    9.258855e-07\n",
      "0.35    1.660467e-06\n",
      "0.59    1.776187e-05\n",
      "0.62    2.530900e-05\n",
      "0.64    3.285487e-05\n",
      "0.65    1.552812e-05\n",
      "0.67    2.710200e-05\n",
      "0.68    4.459380e-06\n",
      "0.69    1.240474e-05\n",
      "0.70    1.953675e-06\n",
      "0.71    3.385328e-06\n",
      "0.72    6.755651e-06\n",
      "0.74    4.666938e-06\n",
      "0.75    6.660624e-06\n",
      "0.76    2.817668e-06\n",
      "0.77    7.394581e-06\n",
      "0.78    7.278298e-06\n",
      "0.79    1.643900e-05\n",
      "0.80    2.359414e-06\n",
      "0.81    7.972868e-06\n",
      "0.82    2.012691e-05\n",
      "0.83    1.285612e-05\n",
      "0.84    2.714514e-06\n",
      "0.86    8.548030e-06\n",
      "0.87    2.154981e-06\n",
      "0.88    8.433623e-07\n",
      "0.90    3.054359e-05\n",
      "0.91    7.735301e-06\n",
      "0.92    2.105467e-05\n",
      "0.93    6.508706e-06\n",
      "0.94    1.698290e-05\n",
      "0.95    1.256479e-05\n",
      "0.96    5.413086e-05\n",
      "0.97    7.686600e-05\n",
      "0.98    6.220625e-05\n",
      "0.99    6.367041e-05\n",
      "Name: query_length, dtype: float64\n",
      "\n",
      "\n",
      "Organization: TGA4\n",
      "identity\n",
      "0.17    0.000003\n",
      "0.22    0.000005\n",
      "0.25    0.000044\n",
      "0.26    0.000027\n",
      "0.27    0.000058\n",
      "          ...   \n",
      "0.96    0.058933\n",
      "0.97    0.040489\n",
      "0.98    0.024802\n",
      "0.99    0.044629\n",
      "1.00    0.011460\n",
      "Name: query_length, Length: 78, dtype: float64\n",
      "\n",
      "\n",
      "Organization: TGVEG\n",
      "identity\n",
      "0.14    0.000011\n",
      "0.16    0.000006\n",
      "0.17    0.000009\n",
      "0.18    0.000009\n",
      "0.19    0.000009\n",
      "          ...   \n",
      "0.96    0.063390\n",
      "0.97    0.065459\n",
      "0.98    0.065561\n",
      "0.99    0.052029\n",
      "1.00    0.009853\n",
      "Name: query_length, Length: 85, dtype: float64\n",
      "\n",
      "\n",
      "Organization: TgRH_pasteur\n",
      "identity\n",
      "0.12    0.000006\n",
      "0.13    0.000019\n",
      "0.16    0.000018\n",
      "0.19    0.000015\n",
      "0.22    0.000006\n",
      "          ...   \n",
      "0.96    0.024590\n",
      "0.97    0.058854\n",
      "0.98    0.155394\n",
      "0.99    0.423077\n",
      "1.00    0.228503\n",
      "Name: query_length, Length: 83, dtype: float64\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Assuming unique_org_names and group_org_idn_seq_len are already defined\n",
    "\n",
    "for org in unique_org_names:\n",
    "    print(f\"Organization: {org}\")\n",
    "    \n",
    "    # Access the data corresponding to the specific org_name\n",
    "    org_data = group_org_idn_seq_len.loc[org]\n",
    "\n",
    "    \n",
    "    print(org_data/total_len)\n",
    "    print(\"\\n\")  # To add a newline between each organization's output\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "org_name\n",
       "TgRH_pasteur           0.979596\n",
       "TGVEG                  0.918427\n",
       "TGA4                   0.910011\n",
       "Chlorocebus_sabaeus    0.817652\n",
       "Name: identity, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paf_max_iden_df.groupby('org_name')['identity'].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x75418455d4d0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paf_max_iden_df.groupby('query_name')[]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>per_matching_residues</th>\n",
       "      <th>query_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96175</th>\n",
       "      <td>0.617884</td>\n",
       "      <td>39550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13567</th>\n",
       "      <td>0.579000</td>\n",
       "      <td>28859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9761</th>\n",
       "      <td>0.568853</td>\n",
       "      <td>27363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13595</th>\n",
       "      <td>0.624966</td>\n",
       "      <td>24838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35963</th>\n",
       "      <td>0.656961</td>\n",
       "      <td>19842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34556</th>\n",
       "      <td>0.938646</td>\n",
       "      <td>1407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>0.856560</td>\n",
       "      <td>1356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14973</th>\n",
       "      <td>0.865160</td>\n",
       "      <td>1349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96106</th>\n",
       "      <td>0.313707</td>\n",
       "      <td>1181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95862</th>\n",
       "      <td>0.608824</td>\n",
       "      <td>1048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       per_matching_residues  query_length\n",
       "96175               0.617884         39550\n",
       "13567               0.579000         28859\n",
       "9761                0.568853         27363\n",
       "13595               0.624966         24838\n",
       "35963               0.656961         19842\n",
       "...                      ...           ...\n",
       "34556               0.938646          1407\n",
       "349                 0.856560          1356\n",
       "14973               0.865160          1349\n",
       "96106               0.313707          1181\n",
       "95862               0.608824          1048\n",
       "\n",
       "[115 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['per_matching_residues'] = df['num_matching_residues']/df['alignment_block_length']\n",
    "\n",
    "df[['per_matching_residues', 'query_length']].sort_values(by=['query_length','per_matching_residues'], ascending=False)\n",
    "#df[['per_matching_residues', 'query_length']].sort_values(by=['per_matching_residues'], ascending=False)\n",
    "#df['mapping_quality'].sort_values( ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "OE1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
